{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Flux Balance Analysis (08302023 data)\n",
    "The purpose of the notebook is to run flux balance analysis to find the set of genome scale fluxes that maximizes biomass production.<br><br>\n",
    "The flux balance analysis results are then compared to reaction rates determined by 13C-metabolic flux analysis (13C-MFA). <br><br>\n",
    "The results of this notebook will be compared to transcript constained genome scale model flux results via E-Flux2. <br><br>\n",
    "This notebook looks at glucose, glycerol, acetate, and oleic acid as sole carbon sources. <br><br>\n",
    "In all cases, parsimonious flux balance analysis (pFBA) was used to prevent degenerate solutions.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "source_dir = '../src'\n",
    "sys.path.append(source_dir)\n",
    "from add_column_to_13c_flux_df import add_column_to_13c_flux_df\n",
    "from flux_prediction_scatterplot import flux_prediction_scatterplot"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the genome scale model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Name</strong></td>\n",
       "                <td>model</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>11156ed10</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of metabolites</strong></td>\n",
       "                <td>1121</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of reactions</strong></td>\n",
       "                <td>1348</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of genes</strong></td>\n",
       "                <td>648</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Number of groups</strong></td>\n",
       "                <td>0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Objective expression</strong></td>\n",
       "                <td>1.0*biomass_C - 1.0*biomass_C_reverse_c1d5c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartments</strong></td>\n",
       "                <td>c, e, m, n, x, r, g, v</td>\n",
       "            </tr>\n",
       "          </table>"
      ],
      "text/plain": [
       "<Model model at 0x11156ed10>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model = cobra.io.json.load_json_model(\"../genome_scale_models/iYLI647_corr_2.json\")\n",
    "model"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load 13C-MFA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 43 reactions in the 13C MFA that are mapped to the GSM\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ID</th>\n",
       "      <th>Equation</th>\n",
       "      <th>reaction_ids</th>\n",
       "      <th>pathway</th>\n",
       "      <th>compartment</th>\n",
       "      <th>glucose</th>\n",
       "      <th>glycerol</th>\n",
       "      <th>oleic acid</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>uptake</td>\n",
       "      <td>Glucose + ATP -&gt; G6P</td>\n",
       "      <td>reverse_EX_glc(e)</td>\n",
       "      <td>substrate_uptake</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>100 ± 0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R3 glyc3p</td>\n",
       "      <td>GLYC + ATP -&gt; Glyc3P</td>\n",
       "      <td>reverse_GLYCt</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100 ± 0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>OA uptake</td>\n",
       "      <td>OA + ATP -&gt; 9*ACCOAcyt + 7*NADH + 7*FADH2</td>\n",
       "      <td>OCDCEAt</td>\n",
       "      <td>substrate_uptake</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100 ± 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>glycolysis/\\ngluconeogensis</td>\n",
       "      <td>R4 net</td>\n",
       "      <td>G6P &lt;-&gt; F6P</td>\n",
       "      <td>PGI</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>17.8 ± 7.9</td>\n",
       "      <td>-12.3 ± 21.2</td>\n",
       "      <td>-151.7 ± 164.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R5 net</td>\n",
       "      <td>F6P + ATP &lt;-&gt; FBP</td>\n",
       "      <td>PFK or reverse_FBP</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>68.1 ± 2.6</td>\n",
       "      <td>-6.7 ± 6.9</td>\n",
       "      <td>-62.8 ± 53.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Unnamed: 0         ID  \\\n",
       "0                          NaN     uptake   \n",
       "1                          NaN  R3 glyc3p   \n",
       "3                          NaN  OA uptake   \n",
       "4  glycolysis/\\ngluconeogensis     R4 net   \n",
       "5                          NaN     R5 net   \n",
       "\n",
       "                                    Equation        reaction_ids  \\\n",
       "0                       Glucose + ATP -> G6P   reverse_EX_glc(e)   \n",
       "1                       GLYC + ATP -> Glyc3P       reverse_GLYCt   \n",
       "3  OA + ATP -> 9*ACCOAcyt + 7*NADH + 7*FADH2             OCDCEAt   \n",
       "4                                G6P <-> F6P                 PGI   \n",
       "5                          F6P + ATP <-> FBP  PFK or reverse_FBP   \n",
       "\n",
       "            pathway compartment     glucose      glycerol      oleic acid  \n",
       "0  substrate_uptake     cytosol     100 ± 0           NaN             NaN  \n",
       "1               emp     cytosol         NaN       100 ± 0             NaN  \n",
       "3  substrate_uptake     cytosol         NaN           NaN         100 ± 0  \n",
       "4               emp     cytosol  17.8 ± 7.9  -12.3 ± 21.2  -151.7 ± 164.5  \n",
       "5               emp     cytosol  68.1 ± 2.6    -6.7 ± 6.9    -62.8 ± 53.4  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load glucose 13C MFA data from ../data/13c_mfa/INCA_model_08302023_GR.xlsx\n",
    "central_rxn_df = pd.read_excel('../data/13c_mfa/INCA_model_08302023_GR.xlsx', sheet_name='0830')\n",
    "\n",
    "# drop every column after 'oleic acid'\n",
    "central_rxn_df = central_rxn_df.loc[:, :'oleic acid']\n",
    "\n",
    "# only keep the reactions that are mapped to the GSM\n",
    "central_rxn_df.dropna(subset = [\"reaction_ids\"], inplace=True)\n",
    "\n",
    "print(f'There are {len(central_rxn_df)} reactions in the 13C MFA that are mapped to the GSM')\n",
    "\n",
    "central_rxn_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate glucose GSM pFBA solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100.0 EX_glc(e)\n",
      "10000.0 EX_h2o(e)\n",
      "10000.0 EX_h(e)\n",
      "10000.0 EX_nh4(e)\n",
      "10000.0 EX_o2(e)\n",
      "10000.0 EX_pi(e)\n",
      "10000.0 EX_so4(e)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>reaction_name</th>\n",
       "      <th>full_reaction</th>\n",
       "      <th>flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>689</th>\n",
       "      <td>H2Otm</td>\n",
       "      <td>H2O transport mitochondrial</td>\n",
       "      <td>h2o[c] &lt;=&gt; h2o[m]</td>\n",
       "      <td>-561.830307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>ATPtm_H</td>\n",
       "      <td>ADPATP transporter mitochondrial</td>\n",
       "      <td>adp[c] + atp[m] + h[c] --&gt; adp[m] + atp[c] + h[m]</td>\n",
       "      <td>341.609652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>PIt2m</td>\n",
       "      <td>phosphate transporter mitochondrial</td>\n",
       "      <td>h[c] + pi[c] &lt;=&gt; h[m] + pi[m]</td>\n",
       "      <td>335.476108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>ATPS3m</td>\n",
       "      <td>ATP synthase mitochondrial</td>\n",
       "      <td>adp[m] + 3.0 h[c] + pi[m] --&gt; atp[m] + h2o[m] ...</td>\n",
       "      <td>326.533474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>959</th>\n",
       "      <td>H2Ot</td>\n",
       "      <td>H2O transport via diffusion</td>\n",
       "      <td>h2o[e] &lt;=&gt; h2o[c]</td>\n",
       "      <td>-325.265198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reaction_id                        reaction_name  \\\n",
       "689       H2Otm          H2O transport mitochondrial   \n",
       "990     ATPtm_H     ADPATP transporter mitochondrial   \n",
       "976       PIt2m  phosphate transporter mitochondrial   \n",
       "607      ATPS3m           ATP synthase mitochondrial   \n",
       "959        H2Ot          H2O transport via diffusion   \n",
       "\n",
       "                                         full_reaction        flux  \n",
       "689                                  h2o[c] <=> h2o[m] -561.830307  \n",
       "990  adp[c] + atp[m] + h[c] --> adp[m] + atp[c] + h[m]  341.609652  \n",
       "976                      h[c] + pi[c] <=> h[m] + pi[m]  335.476108  \n",
       "607  adp[m] + 3.0 h[c] + pi[m] --> atp[m] + h2o[m] ...  326.533474  \n",
       "959                                  h2o[e] <=> h2o[c] -325.265198  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# minimal medium with glucose as sole carbon source\n",
    "model.medium = {\n",
    "  'EX_glc(e)': 100.0, # 100 mmol of glucose\n",
    "  'EX_h2o(e)': 10000.0,\n",
    "  'EX_h(e)': 10000.0,\n",
    "  'EX_nh4(e)': 10000.0,\n",
    "  'EX_o2(e)': 10000.0,\n",
    "  'EX_pi(e)': 10000.0,\n",
    "  'EX_so4(e)': 10000.0,\n",
    "}\n",
    "  \n",
    "with model:\n",
    "  # update the model to use the specified medium\n",
    "  medium = model.medium\n",
    "  [print(model.medium[m], m) for m in model.medium]\n",
    "\n",
    "  # find the optimal solution\n",
    "  glucose_fba_solution = cobra.flux_analysis.pfba(model)\n",
    "\n",
    "  # make a list of dictionaries with the reaction id, name, flux, and absolute flux\n",
    "  reactions = []\n",
    "  for reaction_id, flux in glucose_fba_solution.fluxes.items():\n",
    "    reactions.append({\n",
    "      'reaction_id': reaction_id,\n",
    "      'reaction_name': model.reactions.get_by_id(reaction_id).name,\n",
    "      'full_reaction': model.reactions.get_by_id(reaction_id).reaction,\n",
    "      'flux': flux,\n",
    "      'absolute_flux': abs(flux), # use for sorting, then drop\n",
    "    })\n",
    "\n",
    "  # make a dataframe from the list of dictionaries\n",
    "  glucose_gsm_fba_df = pd.DataFrame(reactions)\n",
    "\n",
    "  # sort the dataframe by absolute flux\n",
    "  glucose_gsm_fba_df = glucose_gsm_fba_df.sort_values(by=['absolute_flux'], ascending=False)\n",
    "\n",
    "  # drop the absolute flux column\n",
    "  glucose_gsm_fba_df = glucose_gsm_fba_df.drop(columns=['absolute_flux'])\n",
    "\n",
    "  # save the dataframe to a csv file\n",
    "  glucose_gsm_fba_df.to_csv('../results/fba_gsm_fluxes/glucose_gsm_fba.csv', index=False)\n",
    "\n",
    "glucose_gsm_fba_df.head()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add glucose pFBA column to 13C-MFA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ID</th>\n",
       "      <th>Equation</th>\n",
       "      <th>reaction_ids</th>\n",
       "      <th>pathway</th>\n",
       "      <th>compartment</th>\n",
       "      <th>glucose</th>\n",
       "      <th>glycerol</th>\n",
       "      <th>oleic acid</th>\n",
       "      <th>glucose pFBA Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>uptake</td>\n",
       "      <td>Glucose + ATP -&gt; G6P</td>\n",
       "      <td>reverse_EX_glc(e)</td>\n",
       "      <td>substrate_uptake</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>100 ± 0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R3 glyc3p</td>\n",
       "      <td>GLYC + ATP -&gt; Glyc3P</td>\n",
       "      <td>reverse_GLYCt</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100 ± 0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>OA uptake</td>\n",
       "      <td>OA + ATP -&gt; 9*ACCOAcyt + 7*NADH + 7*FADH2</td>\n",
       "      <td>OCDCEAt</td>\n",
       "      <td>substrate_uptake</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100 ± 0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>glycolysis/\\ngluconeogensis</td>\n",
       "      <td>R4 net</td>\n",
       "      <td>G6P &lt;-&gt; F6P</td>\n",
       "      <td>PGI</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>17.8 ± 7.9</td>\n",
       "      <td>-12.3 ± 21.2</td>\n",
       "      <td>-151.7 ± 164.5</td>\n",
       "      <td>35.574485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R5 net</td>\n",
       "      <td>F6P + ATP &lt;-&gt; FBP</td>\n",
       "      <td>PFK or reverse_FBP</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>68.1 ± 2.6</td>\n",
       "      <td>-6.7 ± 6.9</td>\n",
       "      <td>-62.8 ± 53.4</td>\n",
       "      <td>38.629379</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Unnamed: 0         ID  \\\n",
       "0                          NaN     uptake   \n",
       "1                          NaN  R3 glyc3p   \n",
       "3                          NaN  OA uptake   \n",
       "4  glycolysis/\\ngluconeogensis     R4 net   \n",
       "5                          NaN     R5 net   \n",
       "\n",
       "                                    Equation        reaction_ids  \\\n",
       "0                       Glucose + ATP -> G6P   reverse_EX_glc(e)   \n",
       "1                       GLYC + ATP -> Glyc3P       reverse_GLYCt   \n",
       "3  OA + ATP -> 9*ACCOAcyt + 7*NADH + 7*FADH2             OCDCEAt   \n",
       "4                                G6P <-> F6P                 PGI   \n",
       "5                          F6P + ATP <-> FBP  PFK or reverse_FBP   \n",
       "\n",
       "            pathway compartment     glucose      glycerol      oleic acid  \\\n",
       "0  substrate_uptake     cytosol     100 ± 0           NaN             NaN   \n",
       "1               emp     cytosol         NaN       100 ± 0             NaN   \n",
       "3  substrate_uptake     cytosol         NaN           NaN         100 ± 0   \n",
       "4               emp     cytosol  17.8 ± 7.9  -12.3 ± 21.2  -151.7 ± 164.5   \n",
       "5               emp     cytosol  68.1 ± 2.6    -6.7 ± 6.9    -62.8 ± 53.4   \n",
       "\n",
       "   glucose pFBA Flux  \n",
       "0         100.000000  \n",
       "1           0.000000  \n",
       "3           0.000000  \n",
       "4          35.574485  \n",
       "5          38.629379  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add the GSM flux predictions to the 13C-MFA dataframe\n",
    "central_rxn_df = add_column_to_13c_flux_df(central_rxn_df, glucose_gsm_fba_df, 'glucose pFBA Flux')\n",
    "\n",
    "central_rxn_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate glycerol GSM pFBA solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100.0 EX_glyc(e)\n",
      "10000.0 EX_h2o(e)\n",
      "10000.0 EX_h(e)\n",
      "10000.0 EX_nh4(e)\n",
      "10000.0 EX_o2(e)\n",
      "10000.0 EX_pi(e)\n",
      "10000.0 EX_so4(e)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>reaction_name</th>\n",
       "      <th>full_reaction</th>\n",
       "      <th>flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>689</th>\n",
       "      <td>H2Otm</td>\n",
       "      <td>H2O transport mitochondrial</td>\n",
       "      <td>h2o[c] &lt;=&gt; h2o[m]</td>\n",
       "      <td>-574.850423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>421</th>\n",
       "      <td>EX_h2o(e)</td>\n",
       "      <td>H2O exchange</td>\n",
       "      <td>h2o[e] &lt;=&gt;</td>\n",
       "      <td>354.104943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>959</th>\n",
       "      <td>H2Ot</td>\n",
       "      <td>H2O transport via diffusion</td>\n",
       "      <td>h2o[e] &lt;=&gt; h2o[c]</td>\n",
       "      <td>-354.104943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>PIt2m</td>\n",
       "      <td>phosphate transporter mitochondrial</td>\n",
       "      <td>h[c] + pi[c] &lt;=&gt; h[m] + pi[m]</td>\n",
       "      <td>315.462030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>ATPS3m</td>\n",
       "      <td>ATP synthase mitochondrial</td>\n",
       "      <td>adp[m] + 3.0 h[c] + pi[m] --&gt; atp[m] + h2o[m] ...</td>\n",
       "      <td>313.155793</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reaction_id                        reaction_name  \\\n",
       "689       H2Otm          H2O transport mitochondrial   \n",
       "421   EX_h2o(e)                         H2O exchange   \n",
       "959        H2Ot          H2O transport via diffusion   \n",
       "976       PIt2m  phosphate transporter mitochondrial   \n",
       "607      ATPS3m           ATP synthase mitochondrial   \n",
       "\n",
       "                                         full_reaction        flux  \n",
       "689                                  h2o[c] <=> h2o[m] -574.850423  \n",
       "421                                        h2o[e] <=>   354.104943  \n",
       "959                                  h2o[e] <=> h2o[c] -354.104943  \n",
       "976                      h[c] + pi[c] <=> h[m] + pi[m]  315.462030  \n",
       "607  adp[m] + 3.0 h[c] + pi[m] --> atp[m] + h2o[m] ...  313.155793  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# minimal medium with glycerol as sole carbon source\n",
    "model.medium = {\n",
    "  'EX_glyc(e)': 100.0, # 100 mmol of glycerol\n",
    "  'EX_h2o(e)': 10000.0,\n",
    "  'EX_h(e)': 10000.0,\n",
    "  'EX_nh4(e)': 10000.0,\n",
    "  'EX_o2(e)': 10000.0,\n",
    "  'EX_pi(e)': 10000.0,\n",
    "  'EX_so4(e)': 10000.0,\n",
    "}\n",
    "  \n",
    "with model:\n",
    "  # update the model to use the specified medium\n",
    "  medium = model.medium\n",
    "  [print(model.medium[m], m) for m in model.medium]\n",
    "\n",
    "  # find the optimal solution\n",
    "  glycerol_fba_solution = cobra.flux_analysis.pfba(model)\n",
    "\n",
    "  # make a list of dictionaries with the reaction id, name, flux, and absolute flux\n",
    "  reactions = []\n",
    "  for reaction_id, flux in glycerol_fba_solution.fluxes.items():\n",
    "    reactions.append({\n",
    "      'reaction_id': reaction_id,\n",
    "      'reaction_name': model.reactions.get_by_id(reaction_id).name,\n",
    "      'full_reaction': model.reactions.get_by_id(reaction_id).reaction,\n",
    "      'flux': flux,\n",
    "      'absolute_flux': abs(flux), # use for sorting, then drop\n",
    "    })\n",
    "\n",
    "  # make a dataframe from the list of dictionaries\n",
    "  glycerol_gsm_fba_df = pd.DataFrame(reactions)\n",
    "\n",
    "  # sort the dataframe by absolute flux\n",
    "  glycerol_gsm_fba_df = glycerol_gsm_fba_df.sort_values(by=['absolute_flux'], ascending=False)\n",
    "\n",
    "  # drop the absolute flux column\n",
    "  glycerol_gsm_fba_df = glycerol_gsm_fba_df.drop(columns=['absolute_flux'])\n",
    "\n",
    "  # save the dataframe to a csv file\n",
    "  glycerol_gsm_fba_df.to_csv('../results/fba_gsm_fluxes/glycerol_gsm_fba.csv', index=False)\n",
    "\n",
    "glycerol_gsm_fba_df.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add glycerol pFBA column to 13C-MFA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ID</th>\n",
       "      <th>Equation</th>\n",
       "      <th>reaction_ids</th>\n",
       "      <th>pathway</th>\n",
       "      <th>compartment</th>\n",
       "      <th>glucose</th>\n",
       "      <th>glycerol</th>\n",
       "      <th>oleic acid</th>\n",
       "      <th>glucose pFBA Flux</th>\n",
       "      <th>glycerol pFBA Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>uptake</td>\n",
       "      <td>Glucose + ATP -&gt; G6P</td>\n",
       "      <td>reverse_EX_glc(e)</td>\n",
       "      <td>substrate_uptake</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>100 ± 0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R3 glyc3p</td>\n",
       "      <td>GLYC + ATP -&gt; Glyc3P</td>\n",
       "      <td>reverse_GLYCt</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100 ± 0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>OA uptake</td>\n",
       "      <td>OA + ATP -&gt; 9*ACCOAcyt + 7*NADH + 7*FADH2</td>\n",
       "      <td>OCDCEAt</td>\n",
       "      <td>substrate_uptake</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100 ± 0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>glycolysis/\\ngluconeogensis</td>\n",
       "      <td>R4 net</td>\n",
       "      <td>G6P &lt;-&gt; F6P</td>\n",
       "      <td>PGI</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>17.8 ± 7.9</td>\n",
       "      <td>-12.3 ± 21.2</td>\n",
       "      <td>-151.7 ± 164.5</td>\n",
       "      <td>35.574485</td>\n",
       "      <td>-27.963948</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R5 net</td>\n",
       "      <td>F6P + ATP &lt;-&gt; FBP</td>\n",
       "      <td>PFK or reverse_FBP</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>68.1 ± 2.6</td>\n",
       "      <td>-6.7 ± 6.9</td>\n",
       "      <td>-62.8 ± 53.4</td>\n",
       "      <td>38.629379</td>\n",
       "      <td>-18.670033</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Unnamed: 0         ID  \\\n",
       "0                          NaN     uptake   \n",
       "1                          NaN  R3 glyc3p   \n",
       "3                          NaN  OA uptake   \n",
       "4  glycolysis/\\ngluconeogensis     R4 net   \n",
       "5                          NaN     R5 net   \n",
       "\n",
       "                                    Equation        reaction_ids  \\\n",
       "0                       Glucose + ATP -> G6P   reverse_EX_glc(e)   \n",
       "1                       GLYC + ATP -> Glyc3P       reverse_GLYCt   \n",
       "3  OA + ATP -> 9*ACCOAcyt + 7*NADH + 7*FADH2             OCDCEAt   \n",
       "4                                G6P <-> F6P                 PGI   \n",
       "5                          F6P + ATP <-> FBP  PFK or reverse_FBP   \n",
       "\n",
       "            pathway compartment     glucose      glycerol      oleic acid  \\\n",
       "0  substrate_uptake     cytosol     100 ± 0           NaN             NaN   \n",
       "1               emp     cytosol         NaN       100 ± 0             NaN   \n",
       "3  substrate_uptake     cytosol         NaN           NaN         100 ± 0   \n",
       "4               emp     cytosol  17.8 ± 7.9  -12.3 ± 21.2  -151.7 ± 164.5   \n",
       "5               emp     cytosol  68.1 ± 2.6    -6.7 ± 6.9    -62.8 ± 53.4   \n",
       "\n",
       "   glucose pFBA Flux  glycerol pFBA Flux  \n",
       "0         100.000000            0.000000  \n",
       "1           0.000000          100.000000  \n",
       "3           0.000000            0.000000  \n",
       "4          35.574485          -27.963948  \n",
       "5          38.629379          -18.670033  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add the GSM flux predictions to the 13C-MFA dataframe\n",
    "central_rxn_df = add_column_to_13c_flux_df(central_rxn_df, glycerol_gsm_fba_df, 'glycerol pFBA Flux')\n",
    "\n",
    "central_rxn_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate oleic acid GSM pFBA solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10000.0 EX_h2o(e)\n",
      "10000.0 EX_h(e)\n",
      "10000.0 EX_nh4(e)\n",
      "10000.0 EX_o2(e)\n",
      "10.0 EX_ocdcea(e)\n",
      "10000.0 EX_pi(e)\n",
      "10000.0 EX_so4(e)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reaction_id</th>\n",
       "      <th>reaction_name</th>\n",
       "      <th>full_reaction</th>\n",
       "      <th>flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>689</th>\n",
       "      <td>H2Otm</td>\n",
       "      <td>H2O transport mitochondrial</td>\n",
       "      <td>h2o[c] &lt;=&gt; h2o[m]</td>\n",
       "      <td>-3584.266821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>607</th>\n",
       "      <td>ATPS3m</td>\n",
       "      <td>ATP synthase mitochondrial</td>\n",
       "      <td>adp[m] + 3.0 h[c] + pi[m] --&gt; atp[m] + h2o[m] ...</td>\n",
       "      <td>2269.151708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>990</th>\n",
       "      <td>ATPtm_H</td>\n",
       "      <td>ADPATP transporter mitochondrial</td>\n",
       "      <td>adp[c] + atp[m] + h[c] --&gt; adp[m] + atp[c] + h[m]</td>\n",
       "      <td>2265.438412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>976</th>\n",
       "      <td>PIt2m</td>\n",
       "      <td>phosphate transporter mitochondrial</td>\n",
       "      <td>h[c] + pi[c] &lt;=&gt; h[m] + pi[m]</td>\n",
       "      <td>2265.438412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>617</th>\n",
       "      <td>CYOR_u6m</td>\n",
       "      <td>ubiquinol 6 cytochrome c reductase</td>\n",
       "      <td>2.0 ficytc[m] + 1.5 h[m] + q6h2[m] --&gt; 2.0 foc...</td>\n",
       "      <td>2053.128358</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    reaction_id                        reaction_name  \\\n",
       "689       H2Otm          H2O transport mitochondrial   \n",
       "607      ATPS3m           ATP synthase mitochondrial   \n",
       "990     ATPtm_H     ADPATP transporter mitochondrial   \n",
       "976       PIt2m  phosphate transporter mitochondrial   \n",
       "617    CYOR_u6m   ubiquinol 6 cytochrome c reductase   \n",
       "\n",
       "                                         full_reaction         flux  \n",
       "689                                  h2o[c] <=> h2o[m] -3584.266821  \n",
       "607  adp[m] + 3.0 h[c] + pi[m] --> atp[m] + h2o[m] ...  2269.151708  \n",
       "990  adp[c] + atp[m] + h[c] --> adp[m] + atp[c] + h[m]  2265.438412  \n",
       "976                      h[c] + pi[c] <=> h[m] + pi[m]  2265.438412  \n",
       "617  2.0 ficytc[m] + 1.5 h[m] + q6h2[m] --> 2.0 foc...  2053.128358  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# minimal medium with oleic acid as sole carbon source\n",
    "model.medium = {\n",
    "  'EX_ocdcea(e)': 10.0, # 10 mmol of oleic acid (to prevent maxing out reactions with 1000 max flux)\n",
    "  'EX_h2o(e)': 10000.0,\n",
    "  'EX_h(e)': 10000.0,\n",
    "  'EX_nh4(e)': 10000.0,\n",
    "  'EX_o2(e)': 10000.0,\n",
    "  'EX_pi(e)': 10000.0,\n",
    "  'EX_so4(e)': 10000.0,\n",
    "}\n",
    "\n",
    "with model:\n",
    "  # update the model to use the specified medium\n",
    "  medium = model.medium\n",
    "  [print(model.medium[m], m) for m in model.medium]\n",
    "\n",
    "  # find the optimal solution\n",
    "  acetate_fba_solution = cobra.flux_analysis.pfba(model)\n",
    "\n",
    "  # make a list of dictionaries with the reaction id, name, flux, and absolute flux\n",
    "  reactions = []\n",
    "  for reaction_id, flux in acetate_fba_solution.fluxes.items():\n",
    "    reactions.append({\n",
    "      'reaction_id': reaction_id,\n",
    "      'reaction_name': model.reactions.get_by_id(reaction_id).name,\n",
    "      'full_reaction': model.reactions.get_by_id(reaction_id).reaction,\n",
    "      'flux': flux * 10, # flux normalized to 100 mols of oleic acid\n",
    "      'absolute_flux': abs(flux), # use for sorting, then drop\n",
    "    })\n",
    "\n",
    "  # make a dataframe from the list of dictionaries\n",
    "  oleic_acid_gsm_fba_df = pd.DataFrame(reactions)\n",
    "\n",
    "  # sort the dataframe by absolute flux\n",
    "  oleic_acid_gsm_fba_df = oleic_acid_gsm_fba_df.sort_values(by=['absolute_flux'], ascending=False)\n",
    "\n",
    "  # drop the absolute flux column\n",
    "  oleic_acid_gsm_fba_df = oleic_acid_gsm_fba_df.drop(columns=['absolute_flux'])\n",
    "\n",
    "  # save the dataframe to a csv file\n",
    "  oleic_acid_gsm_fba_df.to_csv('../results/fba_gsm_fluxes/oleic_acid_gsm_fba.csv', index=False)\n",
    "\n",
    "oleic_acid_gsm_fba_df.head()\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add oleic acid pFBA column to 13C-MFA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ID</th>\n",
       "      <th>Equation</th>\n",
       "      <th>reaction_ids</th>\n",
       "      <th>pathway</th>\n",
       "      <th>compartment</th>\n",
       "      <th>glucose</th>\n",
       "      <th>glycerol</th>\n",
       "      <th>oleic acid</th>\n",
       "      <th>glucose pFBA Flux</th>\n",
       "      <th>glycerol pFBA Flux</th>\n",
       "      <th>oleic acid pFBA Flux</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>uptake</td>\n",
       "      <td>Glucose + ATP -&gt; G6P</td>\n",
       "      <td>reverse_EX_glc(e)</td>\n",
       "      <td>substrate_uptake</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>100 ± 0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R3 glyc3p</td>\n",
       "      <td>GLYC + ATP -&gt; Glyc3P</td>\n",
       "      <td>reverse_GLYCt</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100 ± 0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>OA uptake</td>\n",
       "      <td>OA + ATP -&gt; 9*ACCOAcyt + 7*NADH + 7*FADH2</td>\n",
       "      <td>OCDCEAt</td>\n",
       "      <td>substrate_uptake</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100 ± 0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>glycolysis/\\ngluconeogensis</td>\n",
       "      <td>R4 net</td>\n",
       "      <td>G6P &lt;-&gt; F6P</td>\n",
       "      <td>PGI</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>17.8 ± 7.9</td>\n",
       "      <td>-12.3 ± 21.2</td>\n",
       "      <td>-151.7 ± 164.5</td>\n",
       "      <td>35.574485</td>\n",
       "      <td>-27.963948</td>\n",
       "      <td>-30.701311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>R5 net</td>\n",
       "      <td>F6P + ATP &lt;-&gt; FBP</td>\n",
       "      <td>PFK or reverse_FBP</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>68.1 ± 2.6</td>\n",
       "      <td>-6.7 ± 6.9</td>\n",
       "      <td>-62.8 ± 53.4</td>\n",
       "      <td>38.629379</td>\n",
       "      <td>-18.670033</td>\n",
       "      <td>-69.437259</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Unnamed: 0         ID  \\\n",
       "0                          NaN     uptake   \n",
       "1                          NaN  R3 glyc3p   \n",
       "3                          NaN  OA uptake   \n",
       "4  glycolysis/\\ngluconeogensis     R4 net   \n",
       "5                          NaN     R5 net   \n",
       "\n",
       "                                    Equation        reaction_ids  \\\n",
       "0                       Glucose + ATP -> G6P   reverse_EX_glc(e)   \n",
       "1                       GLYC + ATP -> Glyc3P       reverse_GLYCt   \n",
       "3  OA + ATP -> 9*ACCOAcyt + 7*NADH + 7*FADH2             OCDCEAt   \n",
       "4                                G6P <-> F6P                 PGI   \n",
       "5                          F6P + ATP <-> FBP  PFK or reverse_FBP   \n",
       "\n",
       "            pathway compartment     glucose      glycerol      oleic acid  \\\n",
       "0  substrate_uptake     cytosol     100 ± 0           NaN             NaN   \n",
       "1               emp     cytosol         NaN       100 ± 0             NaN   \n",
       "3  substrate_uptake     cytosol         NaN           NaN         100 ± 0   \n",
       "4               emp     cytosol  17.8 ± 7.9  -12.3 ± 21.2  -151.7 ± 164.5   \n",
       "5               emp     cytosol  68.1 ± 2.6    -6.7 ± 6.9    -62.8 ± 53.4   \n",
       "\n",
       "   glucose pFBA Flux  glycerol pFBA Flux  oleic acid pFBA Flux  \n",
       "0         100.000000            0.000000              0.000000  \n",
       "1           0.000000          100.000000              0.000000  \n",
       "3           0.000000            0.000000            100.000000  \n",
       "4          35.574485          -27.963948            -30.701311  \n",
       "5          38.629379          -18.670033            -69.437259  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add the GSM flux predictions to the 13C-MFA dataframe\n",
    "central_rxn_df = add_column_to_13c_flux_df(central_rxn_df, oleic_acid_gsm_fba_df, 'oleic acid pFBA Flux')\n",
    "\n",
    "central_rxn_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save 13C-MFA data with pFBA data added"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the dataframe to a csv file\n",
    "central_rxn_df.to_csv('../results/central_fluxes/pfba.csv', index=False, encoding='utf-8-sig')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
