{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 13C-MFA Constrained Flux Balance Analysis (08302023 data + StrainDesign) \n",
    "The purpose of the notebook is to run flux balance analysis to find the set of genome scale fluxes that maximizes biomass production.<br><br>\n",
    "The flux balance analysis results are then compared to reaction rates determined by 13C-metabolic flux analysis (13C-MFA). <br><br>\n",
    "The results of this notebook will be compared to transcript constained genome scale model flux results via E-Flux2. <br><br>\n",
    "This notebook looks at glucose, glycerol, acetate, and oleic acid as sole carbon sources. <br><br>\n",
    "In all cases, parsimonious flux balance analysis (pFBA) was used to prevent degenerate solutions.\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "import straindesign as sd\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "\n",
    "source_dir = '../src'\n",
    "sys.path.append(source_dir)\n",
    "from add_flux_column_to_13c_flux_df import add_flux_column_to_13c_flux_df\n",
    "from add_fva_columns_to_13c_flux_df import add_fva_columns_to_13c_flux_df\n",
    "from flux_prediction_scatterplot import flux_prediction_scatterplot\n",
    "from make_rxn_constraint_string import make_rxn_constraint_string"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load the genome scale model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = cobra.io.json.load_json_model('../genome_scale_models/iYLI647_corr_3.json')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rename some reactions to remove parentheses\n",
    "This is because parentheses in reaction ids cause problems with StrainDesign"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>EX_glc_e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>D Glucose exchange</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x15c0bbac0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>glc_D[e] <=></p>\n",
       "                    <p style='text-align:right'>D_Glucose <=></p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td>YALI0D01111g or YALI0D18876g or YALI0D00132g or YALI0B01342g or YALI0E23287g or YALI0B00396g or...</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-10.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction EX_glc_e at 0x15c0bbac0>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.get_by_id('EX_glc(e)').id = 'EX_glc_e'\n",
    "model.reactions.get_by_id('EX_glyc(e)').id = 'EX_glyc_e'\n",
    "model.reactions.get_by_id('EX_ocdcea(e)').id = 'EX_ocdcea_e'\n",
    "model.reactions.get_by_id('EX_h2o(e)').id = 'EX_h2o_e'\n",
    "model.reactions.get_by_id('EX_h(e)').id = 'EX_h_e'\n",
    "model.reactions.get_by_id('EX_nh4(e)').id = 'EX_nh4_e'\n",
    "model.reactions.get_by_id('EX_o2(e)').id = 'EX_o2_e'\n",
    "model.reactions.get_by_id('EX_pi(e)').id = 'EX_pi_e'\n",
    "model.reactions.get_by_id('EX_so4(e)').id = 'EX_so4_e'\n",
    "\n",
    "# print an example reaction\n",
    "model.reactions.get_by_id('EX_glc_e')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load 13C-MFA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 43 reactions in the 13C MFA that are mapped to the GSM\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>ID</th>\n",
       "      <th>Equation</th>\n",
       "      <th>reaction_ids</th>\n",
       "      <th>pathway</th>\n",
       "      <th>compartment</th>\n",
       "      <th>glucose_flux</th>\n",
       "      <th>glucose_std_err</th>\n",
       "      <th>glucose_LB</th>\n",
       "      <th>glucose_UB</th>\n",
       "      <th>...</th>\n",
       "      <th>glycerol_flux</th>\n",
       "      <th>glycerol_std_err</th>\n",
       "      <th>glycerol_LB</th>\n",
       "      <th>glycerol_UB</th>\n",
       "      <th>glycerol_ΔB</th>\n",
       "      <th>oleic_acid_flux</th>\n",
       "      <th>oleic_acid_std_err</th>\n",
       "      <th>oleic_acid_LB</th>\n",
       "      <th>oleic_acid_UB</th>\n",
       "      <th>oleic_acid_ΔB</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>uptake</td>\n",
       "      <td>uptake</td>\n",
       "      <td>Glucose + ATP -&gt; G6P</td>\n",
       "      <td>reverse_EX_glc_e</td>\n",
       "      <td>substrate_uptake</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>1.000000e+02</td>\n",
       "      <td>5590.0000</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>uptake</td>\n",
       "      <td>R3 glyc3p</td>\n",
       "      <td>GLYC + ATP -&gt; Glyc3P</td>\n",
       "      <td>reverse_GLYCt</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>5.974500e-11</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>uptake</td>\n",
       "      <td>R3 dhap net</td>\n",
       "      <td>Glyc3P &lt;-&gt; DHAP + UQH2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>5.974500e-11</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>uptake</td>\n",
       "      <td>OA uptake</td>\n",
       "      <td>OA + ATP -&gt; 9*ACCOAcyt + 7*NADH + 7*FADH2</td>\n",
       "      <td>OCDCEAt</td>\n",
       "      <td>substrate_uptake</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000e+02</td>\n",
       "      <td>2.594100e-10</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>glycolysis/gluconeogensis</td>\n",
       "      <td>R4 net</td>\n",
       "      <td>G6P &lt;-&gt; F6P</td>\n",
       "      <td>PGI</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>1.615850e+01</td>\n",
       "      <td>903.3512</td>\n",
       "      <td>7.6380</td>\n",
       "      <td>25.0760</td>\n",
       "      <td>...</td>\n",
       "      <td>-19.168800</td>\n",
       "      <td>1.740200e+00</td>\n",
       "      <td>-26.9472</td>\n",
       "      <td>-15.8469</td>\n",
       "      <td>11.1003</td>\n",
       "      <td>-1.035480e+02</td>\n",
       "      <td>3.191300e+00</td>\n",
       "      <td>-345.1830</td>\n",
       "      <td>-82.7140</td>\n",
       "      <td>262.4690</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>glycolysis/gluconeogensis</td>\n",
       "      <td>R5 net</td>\n",
       "      <td>F6P + ATP &lt;-&gt; FBP</td>\n",
       "      <td>PFK or reverse_FBP</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>5.470010e+01</td>\n",
       "      <td>3057.7000</td>\n",
       "      <td>50.9378</td>\n",
       "      <td>57.0545</td>\n",
       "      <td>...</td>\n",
       "      <td>-16.913700</td>\n",
       "      <td>6.263000e-01</td>\n",
       "      <td>-19.3382</td>\n",
       "      <td>-15.6909</td>\n",
       "      <td>3.6473</td>\n",
       "      <td>-9.801660e+01</td>\n",
       "      <td>3.020800e+00</td>\n",
       "      <td>-167.8457</td>\n",
       "      <td>-86.4527</td>\n",
       "      <td>81.3930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>glycolysis/gluconeogensis</td>\n",
       "      <td>R6 net</td>\n",
       "      <td>FBP &lt;-&gt; DHAP + GAP</td>\n",
       "      <td>FBA</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>5.470010e+01</td>\n",
       "      <td>3057.7000</td>\n",
       "      <td>50.9378</td>\n",
       "      <td>57.0545</td>\n",
       "      <td>...</td>\n",
       "      <td>-16.913700</td>\n",
       "      <td>6.263000e-01</td>\n",
       "      <td>-19.3382</td>\n",
       "      <td>-15.6909</td>\n",
       "      <td>3.6473</td>\n",
       "      <td>-9.801660e+01</td>\n",
       "      <td>3.020800e+00</td>\n",
       "      <td>-167.8457</td>\n",
       "      <td>-86.4527</td>\n",
       "      <td>81.3930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>glycolysis/gluconeogensis</td>\n",
       "      <td>R7 net</td>\n",
       "      <td>DHAP &lt;-&gt; GAP</td>\n",
       "      <td>TPI</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>5.295380e+01</td>\n",
       "      <td>2960.1000</td>\n",
       "      <td>49.0946</td>\n",
       "      <td>55.3190</td>\n",
       "      <td>...</td>\n",
       "      <td>82.027300</td>\n",
       "      <td>6.319000e-01</td>\n",
       "      <td>78.9156</td>\n",
       "      <td>83.2640</td>\n",
       "      <td>4.3484</td>\n",
       "      <td>-1.083464e+02</td>\n",
       "      <td>3.339200e+00</td>\n",
       "      <td>-176.4324</td>\n",
       "      <td>-95.9590</td>\n",
       "      <td>80.4734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>glycolysis/gluconeogensis</td>\n",
       "      <td>R8 net</td>\n",
       "      <td>GAP &lt;-&gt; G3P + ATP + NADH</td>\n",
       "      <td>GAPD</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>1.262018e+02</td>\n",
       "      <td>7054.7000</td>\n",
       "      <td>120.5904</td>\n",
       "      <td>130.0255</td>\n",
       "      <td>...</td>\n",
       "      <td>65.802800</td>\n",
       "      <td>7.356000e-01</td>\n",
       "      <td>63.6523</td>\n",
       "      <td>67.2932</td>\n",
       "      <td>3.6409</td>\n",
       "      <td>-2.051634e+02</td>\n",
       "      <td>6.323000e+00</td>\n",
       "      <td>-256.9114</td>\n",
       "      <td>-184.8529</td>\n",
       "      <td>72.0585</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>glycolysis/gluconeogensis</td>\n",
       "      <td>R9 net</td>\n",
       "      <td>G3P &lt;-&gt; PEP</td>\n",
       "      <td>ENO</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>1.214871e+02</td>\n",
       "      <td>6791.1000</td>\n",
       "      <td>115.6079</td>\n",
       "      <td>125.6591</td>\n",
       "      <td>...</td>\n",
       "      <td>63.201900</td>\n",
       "      <td>7.767000e-01</td>\n",
       "      <td>61.1082</td>\n",
       "      <td>64.7828</td>\n",
       "      <td>3.6746</td>\n",
       "      <td>-2.169470e+02</td>\n",
       "      <td>7.763400e+00</td>\n",
       "      <td>-262.8657</td>\n",
       "      <td>-194.3019</td>\n",
       "      <td>68.5638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>glycolysis/gluconeogensis</td>\n",
       "      <td>R10 net</td>\n",
       "      <td>PEP &lt;-&gt; PYRcyt + ATP</td>\n",
       "      <td>PYK</td>\n",
       "      <td>emp</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>1.172436e+02</td>\n",
       "      <td>6553.8000</td>\n",
       "      <td>111.0774</td>\n",
       "      <td>126.5098</td>\n",
       "      <td>...</td>\n",
       "      <td>60.628300</td>\n",
       "      <td>7.959000e-01</td>\n",
       "      <td>58.5800</td>\n",
       "      <td>62.2551</td>\n",
       "      <td>3.6751</td>\n",
       "      <td>1.412210e+01</td>\n",
       "      <td>1.434390e+01</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>112.2861</td>\n",
       "      <td>112.2861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>glycolysis/gluconeogensis</td>\n",
       "      <td>R32</td>\n",
       "      <td>OAAcyt + ATP -&gt; PEP + CO2</td>\n",
       "      <td>PPCK</td>\n",
       "      <td>anaplerotic</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>9.999800e-08</td>\n",
       "      <td>7.9151</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>9.6176</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>2.580800e-11</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.7782</td>\n",
       "      <td>0.7782</td>\n",
       "      <td>2.421320e+02</td>\n",
       "      <td>1.836350e+01</td>\n",
       "      <td>204.4483</td>\n",
       "      <td>372.5395</td>\n",
       "      <td>168.0912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>pentose phosphate</td>\n",
       "      <td>R11</td>\n",
       "      <td>G6P -&gt; PG6 + NADPH</td>\n",
       "      <td>G6PDH2</td>\n",
       "      <td>pentose_phosphate</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>6.092920e+01</td>\n",
       "      <td>3406.0000</td>\n",
       "      <td>50.2982</td>\n",
       "      <td>69.3906</td>\n",
       "      <td>...</td>\n",
       "      <td>5.273000</td>\n",
       "      <td>1.697800e+00</td>\n",
       "      <td>2.0588</td>\n",
       "      <td>15.8682</td>\n",
       "      <td>13.8094</td>\n",
       "      <td>1.736070e+01</td>\n",
       "      <td>5.350000e-01</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>273.5399</td>\n",
       "      <td>273.5399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>pentose phosphate</td>\n",
       "      <td>R12.2</td>\n",
       "      <td>PG6 -&gt; Ru5P + CO2 + NADPH</td>\n",
       "      <td>GND</td>\n",
       "      <td>pentose_phosphate</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>6.092920e+01</td>\n",
       "      <td>3406.0000</td>\n",
       "      <td>50.2982</td>\n",
       "      <td>69.3906</td>\n",
       "      <td>...</td>\n",
       "      <td>5.273000</td>\n",
       "      <td>1.697800e+00</td>\n",
       "      <td>2.0588</td>\n",
       "      <td>15.8682</td>\n",
       "      <td>13.8094</td>\n",
       "      <td>1.736070e+01</td>\n",
       "      <td>5.350000e-01</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>273.5399</td>\n",
       "      <td>273.5399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>pentose phosphate</td>\n",
       "      <td>R13.2 net</td>\n",
       "      <td>Ru5P &lt;-&gt; R5P</td>\n",
       "      <td>reverse_RPI</td>\n",
       "      <td>pentose_phosphate</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>2.238770e+01</td>\n",
       "      <td>1251.5000</td>\n",
       "      <td>19.0080</td>\n",
       "      <td>25.2153</td>\n",
       "      <td>...</td>\n",
       "      <td>3.017900</td>\n",
       "      <td>5.694000e-01</td>\n",
       "      <td>1.9264</td>\n",
       "      <td>6.5241</td>\n",
       "      <td>4.5977</td>\n",
       "      <td>1.182930e+01</td>\n",
       "      <td>3.646000e-01</td>\n",
       "      <td>5.8507</td>\n",
       "      <td>96.2027</td>\n",
       "      <td>90.3520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>pentose phosphate</td>\n",
       "      <td>R14.2 net</td>\n",
       "      <td>Ru5P &lt;-&gt; X5P</td>\n",
       "      <td>RPE</td>\n",
       "      <td>pentose_phosphate</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>3.854150e+01</td>\n",
       "      <td>2154.5000</td>\n",
       "      <td>31.2322</td>\n",
       "      <td>42.9235</td>\n",
       "      <td>...</td>\n",
       "      <td>2.255100</td>\n",
       "      <td>1.128700e+00</td>\n",
       "      <td>0.1059</td>\n",
       "      <td>7.6108</td>\n",
       "      <td>7.5049</td>\n",
       "      <td>5.531400e+00</td>\n",
       "      <td>1.705000e-01</td>\n",
       "      <td>-6.2991</td>\n",
       "      <td>177.3372</td>\n",
       "      <td>183.6363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>pentose phosphate</td>\n",
       "      <td>R15.2 net</td>\n",
       "      <td>X5P &lt;-&gt; GAP + TKC2</td>\n",
       "      <td>TKT1 or TKT2</td>\n",
       "      <td>pentose_phosphate</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>3.854150e+01</td>\n",
       "      <td>2154.5000</td>\n",
       "      <td>31.2322</td>\n",
       "      <td>42.9235</td>\n",
       "      <td>...</td>\n",
       "      <td>2.255100</td>\n",
       "      <td>1.128700e+00</td>\n",
       "      <td>0.1059</td>\n",
       "      <td>7.6108</td>\n",
       "      <td>7.5049</td>\n",
       "      <td>5.531400e+00</td>\n",
       "      <td>1.705000e-01</td>\n",
       "      <td>-6.2991</td>\n",
       "      <td>177.3372</td>\n",
       "      <td>183.6363</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>pentose phosphate</td>\n",
       "      <td>R16.2 net</td>\n",
       "      <td>E4P + TKC2 &lt;-&gt; F6P</td>\n",
       "      <td>TKT2</td>\n",
       "      <td>pentose_phosphate</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>1.820990e+01</td>\n",
       "      <td>1017.9000</td>\n",
       "      <td>14.4968</td>\n",
       "      <td>21.0188</td>\n",
       "      <td>...</td>\n",
       "      <td>0.484200</td>\n",
       "      <td>5.628000e-01</td>\n",
       "      <td>-0.5867</td>\n",
       "      <td>4.0518</td>\n",
       "      <td>4.6385</td>\n",
       "      <td>-7.205800e-07</td>\n",
       "      <td>3.578900e-09</td>\n",
       "      <td>-6.0327</td>\n",
       "      <td>86.3696</td>\n",
       "      <td>92.4023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>pentose phosphate</td>\n",
       "      <td>R17.2 net</td>\n",
       "      <td>R5P + TKC2 &lt;-&gt; S7P</td>\n",
       "      <td>TKT1</td>\n",
       "      <td>pentose_phosphate</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>2.033160e+01</td>\n",
       "      <td>1136.6000</td>\n",
       "      <td>16.7898</td>\n",
       "      <td>23.1501</td>\n",
       "      <td>...</td>\n",
       "      <td>1.771000</td>\n",
       "      <td>5.660000e-01</td>\n",
       "      <td>0.6921</td>\n",
       "      <td>4.4382</td>\n",
       "      <td>3.7461</td>\n",
       "      <td>5.531400e+00</td>\n",
       "      <td>1.705000e-01</td>\n",
       "      <td>-0.2663</td>\n",
       "      <td>90.9676</td>\n",
       "      <td>91.2339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>pentose phosphate</td>\n",
       "      <td>R18.2 net</td>\n",
       "      <td>GAP + TAC3 &lt;-&gt; F6P</td>\n",
       "      <td>TALA</td>\n",
       "      <td>pentose_phosphate</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>2.033160e+01</td>\n",
       "      <td>1136.6000</td>\n",
       "      <td>16.7898</td>\n",
       "      <td>23.1501</td>\n",
       "      <td>...</td>\n",
       "      <td>1.771000</td>\n",
       "      <td>5.660000e-01</td>\n",
       "      <td>0.6921</td>\n",
       "      <td>4.4382</td>\n",
       "      <td>3.7461</td>\n",
       "      <td>5.531400e+00</td>\n",
       "      <td>1.705000e-01</td>\n",
       "      <td>-0.2663</td>\n",
       "      <td>90.9676</td>\n",
       "      <td>91.2339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>pentose phosphate</td>\n",
       "      <td>R19.2 net</td>\n",
       "      <td>S7P &lt;-&gt; E4P + TAC3</td>\n",
       "      <td>TALA</td>\n",
       "      <td>pentose_phosphate</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>2.033160e+01</td>\n",
       "      <td>1136.6000</td>\n",
       "      <td>16.7898</td>\n",
       "      <td>23.1501</td>\n",
       "      <td>...</td>\n",
       "      <td>1.771000</td>\n",
       "      <td>5.660000e-01</td>\n",
       "      <td>0.6921</td>\n",
       "      <td>4.4382</td>\n",
       "      <td>3.7461</td>\n",
       "      <td>5.531400e+00</td>\n",
       "      <td>1.705000e-01</td>\n",
       "      <td>-0.2663</td>\n",
       "      <td>90.9676</td>\n",
       "      <td>91.2339</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>TCA cycle</td>\n",
       "      <td>R19</td>\n",
       "      <td>PYRmit -&gt; ACCOAmit + CO2 + NADH</td>\n",
       "      <td>PDHm</td>\n",
       "      <td>tca_cycle</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>9.279170e+01</td>\n",
       "      <td>5187.0000</td>\n",
       "      <td>86.6629</td>\n",
       "      <td>100.6986</td>\n",
       "      <td>...</td>\n",
       "      <td>43.592200</td>\n",
       "      <td>9.147000e-01</td>\n",
       "      <td>41.5129</td>\n",
       "      <td>45.4798</td>\n",
       "      <td>3.9669</td>\n",
       "      <td>8.046600e-07</td>\n",
       "      <td>6.512000e-11</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>96.2850</td>\n",
       "      <td>96.2850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>TCA cycle</td>\n",
       "      <td>R20</td>\n",
       "      <td>ACCOAmit + OAAmit -&gt; CITmit</td>\n",
       "      <td>CSm</td>\n",
       "      <td>tca_cycle</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>4.028740e+01</td>\n",
       "      <td>2252.0000</td>\n",
       "      <td>28.1700</td>\n",
       "      <td>51.2108</td>\n",
       "      <td>...</td>\n",
       "      <td>13.956100</td>\n",
       "      <td>1.206500e+00</td>\n",
       "      <td>12.5923</td>\n",
       "      <td>16.5250</td>\n",
       "      <td>3.9327</td>\n",
       "      <td>3.644949e+02</td>\n",
       "      <td>1.751540e+01</td>\n",
       "      <td>347.2815</td>\n",
       "      <td>430.2887</td>\n",
       "      <td>83.0072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>TCA cycle</td>\n",
       "      <td>R21 net</td>\n",
       "      <td>CITmit &lt;-&gt; ICTmit</td>\n",
       "      <td>ACONTm</td>\n",
       "      <td>tca_cycle</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>2.796850e+01</td>\n",
       "      <td>1563.6000</td>\n",
       "      <td>12.6008</td>\n",
       "      <td>40.3555</td>\n",
       "      <td>...</td>\n",
       "      <td>8.174700</td>\n",
       "      <td>1.311100e+00</td>\n",
       "      <td>7.1646</td>\n",
       "      <td>10.9497</td>\n",
       "      <td>3.7851</td>\n",
       "      <td>5.138920e+01</td>\n",
       "      <td>2.615360e+01</td>\n",
       "      <td>21.0202</td>\n",
       "      <td>110.6330</td>\n",
       "      <td>89.6128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TCA cycle</td>\n",
       "      <td>R23</td>\n",
       "      <td>ICTmit -&gt; AKG + CO2 + NADH</td>\n",
       "      <td>ICDHxm</td>\n",
       "      <td>tca_cycle</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>2.796850e+01</td>\n",
       "      <td>89320.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>40.3559</td>\n",
       "      <td>...</td>\n",
       "      <td>4.161800</td>\n",
       "      <td>3.533800e+05</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>10.9497</td>\n",
       "      <td>10.9497</td>\n",
       "      <td>2.058700e-06</td>\n",
       "      <td>3.195200e+05</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>85.2036</td>\n",
       "      <td>85.2036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>TCA cycle</td>\n",
       "      <td>R24</td>\n",
       "      <td>ICTmit -&gt; AKG + CO2 + NADPH</td>\n",
       "      <td>ICDHym</td>\n",
       "      <td>tca_cycle</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>1.000200e-07</td>\n",
       "      <td>89302.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>30.8352</td>\n",
       "      <td>...</td>\n",
       "      <td>4.012900</td>\n",
       "      <td>3.533800e+05</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>9.1309</td>\n",
       "      <td>9.1309</td>\n",
       "      <td>5.138920e+01</td>\n",
       "      <td>3.195200e+05</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>110.6141</td>\n",
       "      <td>110.6141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>TCA cycle</td>\n",
       "      <td>R25</td>\n",
       "      <td>AKG -&gt; 0.5*SUCmit + 0.5*SUCmit + CO2 + ATP + NADH</td>\n",
       "      <td>AKGDam and AKGDbm</td>\n",
       "      <td>tca_cycle</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>1.596080e+01</td>\n",
       "      <td>892.6194</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>29.3331</td>\n",
       "      <td>...</td>\n",
       "      <td>0.892300</td>\n",
       "      <td>1.383800e+00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>3.8277</td>\n",
       "      <td>3.8277</td>\n",
       "      <td>2.509820e+01</td>\n",
       "      <td>2.696390e+01</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>86.2498</td>\n",
       "      <td>86.2498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>TCA cycle</td>\n",
       "      <td>R26 net</td>\n",
       "      <td>SUCmit &lt;-&gt; FUMmit + UQH2</td>\n",
       "      <td>SUCD2_u6m and SUCD1m</td>\n",
       "      <td>tca_cycle</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>2.827970e+01</td>\n",
       "      <td>1580.9000</td>\n",
       "      <td>15.0997</td>\n",
       "      <td>40.2079</td>\n",
       "      <td>...</td>\n",
       "      <td>6.673700</td>\n",
       "      <td>1.277000e+00</td>\n",
       "      <td>5.4743</td>\n",
       "      <td>9.4004</td>\n",
       "      <td>3.9261</td>\n",
       "      <td>3.382039e+02</td>\n",
       "      <td>1.832330e+01</td>\n",
       "      <td>319.8737</td>\n",
       "      <td>410.6453</td>\n",
       "      <td>90.7716</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>TCA cycle</td>\n",
       "      <td>R27 net</td>\n",
       "      <td>0.5*FUMmit + 0.5*FUMmit &lt;-&gt; 0.5*MALmit + 0.5*M...</td>\n",
       "      <td>FUMm</td>\n",
       "      <td>tca_cycle</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>3.006350e+01</td>\n",
       "      <td>1680.6000</td>\n",
       "      <td>17.1305</td>\n",
       "      <td>41.8427</td>\n",
       "      <td>...</td>\n",
       "      <td>7.755500</td>\n",
       "      <td>1.266400e+00</td>\n",
       "      <td>6.5658</td>\n",
       "      <td>10.4589</td>\n",
       "      <td>3.8931</td>\n",
       "      <td>3.418693e+02</td>\n",
       "      <td>1.821070e+01</td>\n",
       "      <td>323.6948</td>\n",
       "      <td>413.3904</td>\n",
       "      <td>89.6956</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>TCA cycle</td>\n",
       "      <td>R28 net</td>\n",
       "      <td>MALmit &lt;-&gt; OAAmit + NADH</td>\n",
       "      <td>MDHm</td>\n",
       "      <td>tca_cycle</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>4.028740e+01</td>\n",
       "      <td>2252.0000</td>\n",
       "      <td>28.1700</td>\n",
       "      <td>51.2108</td>\n",
       "      <td>...</td>\n",
       "      <td>13.956100</td>\n",
       "      <td>1.206500e+00</td>\n",
       "      <td>12.5923</td>\n",
       "      <td>16.5250</td>\n",
       "      <td>3.9327</td>\n",
       "      <td>3.644949e+02</td>\n",
       "      <td>1.751540e+01</td>\n",
       "      <td>347.2815</td>\n",
       "      <td>430.2887</td>\n",
       "      <td>83.0072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>pyruvate metabolism</td>\n",
       "      <td>R29 net</td>\n",
       "      <td>MALmit &lt;-&gt; PYRmit + CO2 + NADH</td>\n",
       "      <td>ME1m</td>\n",
       "      <td>anaplerotic</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>-1.022390e+01</td>\n",
       "      <td>571.4598</td>\n",
       "      <td>-11.2237</td>\n",
       "      <td>-4.7120</td>\n",
       "      <td>...</td>\n",
       "      <td>-6.200600</td>\n",
       "      <td>6.770000e-02</td>\n",
       "      <td>-6.2882</td>\n",
       "      <td>-5.9066</td>\n",
       "      <td>0.3816</td>\n",
       "      <td>2.732400e+01</td>\n",
       "      <td>8.421000e-01</td>\n",
       "      <td>-23.5867</td>\n",
       "      <td>120.2857</td>\n",
       "      <td>143.8724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>pyruvate metabolism</td>\n",
       "      <td>R31</td>\n",
       "      <td>PYRcyt + CO2 + ATP -&gt; OAAcyt</td>\n",
       "      <td>PC</td>\n",
       "      <td>anaplerotic</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>1.328400e+00</td>\n",
       "      <td>74.2421</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>5.7862</td>\n",
       "      <td>...</td>\n",
       "      <td>3.012200</td>\n",
       "      <td>2.388000e-01</td>\n",
       "      <td>2.5256</td>\n",
       "      <td>3.5355</td>\n",
       "      <td>1.0099</td>\n",
       "      <td>1.532170e+01</td>\n",
       "      <td>1.435670e+01</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>62.3378</td>\n",
       "      <td>62.3378</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>pyruvate metabolism</td>\n",
       "      <td>R35</td>\n",
       "      <td>AA -&gt; AC + NADH</td>\n",
       "      <td>ALDD2y</td>\n",
       "      <td>acetate_metabolism</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>1.573800e-05</td>\n",
       "      <td>60003.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.1631</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>2.441600e+05</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.2015</td>\n",
       "      <td>0.2015</td>\n",
       "      <td>9.578000e-01</td>\n",
       "      <td>3.289600e+00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>24.7619</td>\n",
       "      <td>24.7619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>pyruvate metabolism</td>\n",
       "      <td>R36</td>\n",
       "      <td>AA -&gt; AC + NADPH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>acetate_metabolism</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>9.542000e-01</td>\n",
       "      <td>60003.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>2.1173</td>\n",
       "      <td>...</td>\n",
       "      <td>1.095600</td>\n",
       "      <td>2.441600e+05</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.3082</td>\n",
       "      <td>1.3082</td>\n",
       "      <td>1.001600e-07</td>\n",
       "      <td>9.546500e-09</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>23.8039</td>\n",
       "      <td>23.8039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>pyruvate metabolism</td>\n",
       "      <td>R37 net</td>\n",
       "      <td>AC + ATP &lt;-&gt; ACCOAcyt</td>\n",
       "      <td>ACS or reverse_ACOAH</td>\n",
       "      <td>acetate_metabolism</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>2.343700e+00</td>\n",
       "      <td>131.0054</td>\n",
       "      <td>1.3221</td>\n",
       "      <td>3.5411</td>\n",
       "      <td>...</td>\n",
       "      <td>1.938200</td>\n",
       "      <td>9.150000e-02</td>\n",
       "      <td>1.6632</td>\n",
       "      <td>2.1408</td>\n",
       "      <td>0.4776</td>\n",
       "      <td>5.889500e+00</td>\n",
       "      <td>3.204700e+00</td>\n",
       "      <td>3.8578</td>\n",
       "      <td>29.7333</td>\n",
       "      <td>25.8755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>glyoxylate shunt</td>\n",
       "      <td>R41.2 net</td>\n",
       "      <td>CITcyt &lt;-&gt; ICTcyt</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.231890e+01</td>\n",
       "      <td>688.7175</td>\n",
       "      <td>9.9873</td>\n",
       "      <td>20.8181</td>\n",
       "      <td>...</td>\n",
       "      <td>5.781400</td>\n",
       "      <td>2.544000e-01</td>\n",
       "      <td>5.3203</td>\n",
       "      <td>6.3999</td>\n",
       "      <td>1.0796</td>\n",
       "      <td>3.131057e+02</td>\n",
       "      <td>8.793700e+00</td>\n",
       "      <td>291.4722</td>\n",
       "      <td>404.4560</td>\n",
       "      <td>112.9838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>glyoxylate shunt</td>\n",
       "      <td>R43.2</td>\n",
       "      <td>ICTcyt -&gt; GLYOXYcyt + SUCcyt</td>\n",
       "      <td>ICL</td>\n",
       "      <td>glyoxylate_shunt</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>1.231890e+01</td>\n",
       "      <td>688.7175</td>\n",
       "      <td>9.9873</td>\n",
       "      <td>20.8181</td>\n",
       "      <td>...</td>\n",
       "      <td>5.781400</td>\n",
       "      <td>2.544000e-01</td>\n",
       "      <td>5.3203</td>\n",
       "      <td>6.3999</td>\n",
       "      <td>1.0796</td>\n",
       "      <td>3.131057e+02</td>\n",
       "      <td>8.793700e+00</td>\n",
       "      <td>291.4722</td>\n",
       "      <td>404.4560</td>\n",
       "      <td>112.9838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>glyoxylate shunt</td>\n",
       "      <td>R45</td>\n",
       "      <td>ACCOAcyt + GLYOXYcyt -&gt; MALcyt</td>\n",
       "      <td>MALSp</td>\n",
       "      <td>glyoxylate_shunt</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>1.231890e+01</td>\n",
       "      <td>688.7175</td>\n",
       "      <td>9.9873</td>\n",
       "      <td>20.8181</td>\n",
       "      <td>...</td>\n",
       "      <td>5.781400</td>\n",
       "      <td>2.544000e-01</td>\n",
       "      <td>5.3203</td>\n",
       "      <td>6.3999</td>\n",
       "      <td>1.0796</td>\n",
       "      <td>3.131057e+02</td>\n",
       "      <td>8.793700e+00</td>\n",
       "      <td>291.4722</td>\n",
       "      <td>404.4560</td>\n",
       "      <td>112.9838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>glyoxylate shunt</td>\n",
       "      <td>R46</td>\n",
       "      <td>MALcyt -&gt; OAAcyt + NADH</td>\n",
       "      <td>MDH</td>\n",
       "      <td>tca_cycle</td>\n",
       "      <td>cytosol</td>\n",
       "      <td>1.231890e+01</td>\n",
       "      <td>688.5462</td>\n",
       "      <td>9.9871</td>\n",
       "      <td>21.7042</td>\n",
       "      <td>...</td>\n",
       "      <td>5.781400</td>\n",
       "      <td>2.544000e-01</td>\n",
       "      <td>5.3203</td>\n",
       "      <td>6.3997</td>\n",
       "      <td>1.0794</td>\n",
       "      <td>2.631561e+02</td>\n",
       "      <td>7.276300e+00</td>\n",
       "      <td>239.1927</td>\n",
       "      <td>390.4334</td>\n",
       "      <td>151.2407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>compartment transport</td>\n",
       "      <td>R18</td>\n",
       "      <td>PYRcyt -&gt; PYRmit</td>\n",
       "      <td>PYRt2m</td>\n",
       "      <td>transport</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>1.162532e+02</td>\n",
       "      <td>6498.4000</td>\n",
       "      <td>109.8734</td>\n",
       "      <td>123.6819</td>\n",
       "      <td>...</td>\n",
       "      <td>57.821000</td>\n",
       "      <td>8.105000e-01</td>\n",
       "      <td>55.5389</td>\n",
       "      <td>59.4518</td>\n",
       "      <td>3.9129</td>\n",
       "      <td>8.492400e-07</td>\n",
       "      <td>2.092900e-10</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>118.6184</td>\n",
       "      <td>118.6184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>compartment transport</td>\n",
       "      <td>R42 net</td>\n",
       "      <td>ACCOAcyt &lt;-&gt; ACCOAmit</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-5.025110e+01</td>\n",
       "      <td>2809.2000</td>\n",
       "      <td>-61.8069</td>\n",
       "      <td>-45.1953</td>\n",
       "      <td>...</td>\n",
       "      <td>-28.269600</td>\n",
       "      <td>4.550000e-01</td>\n",
       "      <td>-29.0546</td>\n",
       "      <td>-27.2949</td>\n",
       "      <td>1.7597</td>\n",
       "      <td>3.691600e+02</td>\n",
       "      <td>1.737200e+01</td>\n",
       "      <td>316.6482</td>\n",
       "      <td>423.2757</td>\n",
       "      <td>106.6275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>compartment transport</td>\n",
       "      <td>cit transp</td>\n",
       "      <td>CITmit -&gt; CITcyt</td>\n",
       "      <td>CITtam or CITtbm or CITtcm</td>\n",
       "      <td>transport</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>1.231890e+01</td>\n",
       "      <td>688.7175</td>\n",
       "      <td>9.9873</td>\n",
       "      <td>20.8181</td>\n",
       "      <td>...</td>\n",
       "      <td>5.781400</td>\n",
       "      <td>2.544000e-01</td>\n",
       "      <td>5.3203</td>\n",
       "      <td>6.3999</td>\n",
       "      <td>1.0796</td>\n",
       "      <td>3.131057e+02</td>\n",
       "      <td>8.793700e+00</td>\n",
       "      <td>291.4722</td>\n",
       "      <td>404.4560</td>\n",
       "      <td>112.9838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>compartment transport</td>\n",
       "      <td>R44</td>\n",
       "      <td>SUCcyt -&gt; SUCmit</td>\n",
       "      <td>SUCCtm</td>\n",
       "      <td>transport</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>1.231890e+01</td>\n",
       "      <td>688.7175</td>\n",
       "      <td>9.9873</td>\n",
       "      <td>20.8181</td>\n",
       "      <td>...</td>\n",
       "      <td>5.781400</td>\n",
       "      <td>2.544000e-01</td>\n",
       "      <td>5.3203</td>\n",
       "      <td>6.3999</td>\n",
       "      <td>1.0796</td>\n",
       "      <td>3.131057e+02</td>\n",
       "      <td>8.793700e+00</td>\n",
       "      <td>291.4722</td>\n",
       "      <td>404.4560</td>\n",
       "      <td>112.9838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>compartment transport</td>\n",
       "      <td>R47</td>\n",
       "      <td>MALcyt -&gt; MALmit</td>\n",
       "      <td>MALtm or reverse_CITtam or reverse_DICtm</td>\n",
       "      <td>transport</td>\n",
       "      <td>mitochondria</td>\n",
       "      <td>1.000100e-07</td>\n",
       "      <td>9.2808</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>6.1103</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>3.522900e-11</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3057</td>\n",
       "      <td>0.3057</td>\n",
       "      <td>4.994960e+01</td>\n",
       "      <td>1.539400e+00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>141.0555</td>\n",
       "      <td>141.0555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>amino acid biosynthesis</td>\n",
       "      <td>R48</td>\n",
       "      <td>G3P -&gt; SER + NADH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.714600e+00</td>\n",
       "      <td>263.5628</td>\n",
       "      <td>4.0957</td>\n",
       "      <td>5.4976</td>\n",
       "      <td>...</td>\n",
       "      <td>2.600900</td>\n",
       "      <td>5.770000e-02</td>\n",
       "      <td>2.4696</td>\n",
       "      <td>2.7419</td>\n",
       "      <td>0.2723</td>\n",
       "      <td>1.178360e+01</td>\n",
       "      <td>1.887400e+00</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>12.7834</td>\n",
       "      <td>12.7834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>amino acid biosynthesis</td>\n",
       "      <td>R50 net</td>\n",
       "      <td>SER &lt;-&gt; GLY + MTHF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.973200e+00</td>\n",
       "      <td>110.3147</td>\n",
       "      <td>1.4156</td>\n",
       "      <td>2.5653</td>\n",
       "      <td>...</td>\n",
       "      <td>0.938300</td>\n",
       "      <td>4.880000e-02</td>\n",
       "      <td>0.8274</td>\n",
       "      <td>1.0741</td>\n",
       "      <td>0.2467</td>\n",
       "      <td>5.719000e+00</td>\n",
       "      <td>1.761500e+00</td>\n",
       "      <td>-6.0935</td>\n",
       "      <td>6.4612</td>\n",
       "      <td>12.5547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>amino acid biosynthesis</td>\n",
       "      <td>R51 net</td>\n",
       "      <td>CO2 + MTHF + NADH &lt;-&gt; GLY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.585000e-01</td>\n",
       "      <td>20.0475</td>\n",
       "      <td>-0.2078</td>\n",
       "      <td>0.8764</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.041000</td>\n",
       "      <td>4.610000e-02</td>\n",
       "      <td>-0.1422</td>\n",
       "      <td>0.0949</td>\n",
       "      <td>0.2371</td>\n",
       "      <td>9.873000e-01</td>\n",
       "      <td>1.671200e+00</td>\n",
       "      <td>-10.8917</td>\n",
       "      <td>1.5284</td>\n",
       "      <td>12.4201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>amino acid biosynthesis</td>\n",
       "      <td>R52</td>\n",
       "      <td>AKG + NADPH -&gt; GLU</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.058240e+01</td>\n",
       "      <td>2268.6000</td>\n",
       "      <td>36.9632</td>\n",
       "      <td>45.1679</td>\n",
       "      <td>...</td>\n",
       "      <td>25.129200</td>\n",
       "      <td>2.759000e-01</td>\n",
       "      <td>24.5244</td>\n",
       "      <td>25.5919</td>\n",
       "      <td>1.0675</td>\n",
       "      <td>9.096050e+01</td>\n",
       "      <td>2.834100e+00</td>\n",
       "      <td>69.6159</td>\n",
       "      <td>115.4479</td>\n",
       "      <td>45.8320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>amino acid biosynthesis</td>\n",
       "      <td>R53</td>\n",
       "      <td>GLU + ATP -&gt; GLN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.205700e+00</td>\n",
       "      <td>346.9084</td>\n",
       "      <td>5.6858</td>\n",
       "      <td>6.8391</td>\n",
       "      <td>...</td>\n",
       "      <td>3.763600</td>\n",
       "      <td>4.110000e-02</td>\n",
       "      <td>3.6714</td>\n",
       "      <td>3.8168</td>\n",
       "      <td>0.1454</td>\n",
       "      <td>1.936010e+01</td>\n",
       "      <td>5.967000e-01</td>\n",
       "      <td>14.6939</td>\n",
       "      <td>20.1824</td>\n",
       "      <td>5.4885</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>49 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Unnamed: 0           ID  \\\n",
       "0                      uptake       uptake   \n",
       "1                      uptake    R3 glyc3p   \n",
       "2                      uptake  R3 dhap net   \n",
       "3                      uptake    OA uptake   \n",
       "4   glycolysis/gluconeogensis       R4 net   \n",
       "5   glycolysis/gluconeogensis       R5 net   \n",
       "6   glycolysis/gluconeogensis       R6 net   \n",
       "7   glycolysis/gluconeogensis       R7 net   \n",
       "8   glycolysis/gluconeogensis       R8 net   \n",
       "9   glycolysis/gluconeogensis       R9 net   \n",
       "10  glycolysis/gluconeogensis      R10 net   \n",
       "11  glycolysis/gluconeogensis          R32   \n",
       "12          pentose phosphate          R11   \n",
       "13          pentose phosphate        R12.2   \n",
       "14          pentose phosphate    R13.2 net   \n",
       "15          pentose phosphate    R14.2 net   \n",
       "16          pentose phosphate    R15.2 net   \n",
       "17          pentose phosphate    R16.2 net   \n",
       "18          pentose phosphate    R17.2 net   \n",
       "19          pentose phosphate    R18.2 net   \n",
       "20          pentose phosphate    R19.2 net   \n",
       "21                  TCA cycle          R19   \n",
       "22                  TCA cycle          R20   \n",
       "23                  TCA cycle      R21 net   \n",
       "24                  TCA cycle          R23   \n",
       "25                  TCA cycle          R24   \n",
       "26                  TCA cycle          R25   \n",
       "27                  TCA cycle      R26 net   \n",
       "28                  TCA cycle      R27 net   \n",
       "29                  TCA cycle      R28 net   \n",
       "30        pyruvate metabolism      R29 net   \n",
       "31        pyruvate metabolism          R31   \n",
       "32        pyruvate metabolism          R35   \n",
       "33        pyruvate metabolism          R36   \n",
       "34        pyruvate metabolism      R37 net   \n",
       "35           glyoxylate shunt    R41.2 net   \n",
       "36           glyoxylate shunt        R43.2   \n",
       "37           glyoxylate shunt          R45   \n",
       "38           glyoxylate shunt          R46   \n",
       "39      compartment transport          R18   \n",
       "40      compartment transport      R42 net   \n",
       "41      compartment transport   cit transp   \n",
       "42      compartment transport          R44   \n",
       "43      compartment transport          R47   \n",
       "44    amino acid biosynthesis          R48   \n",
       "45    amino acid biosynthesis      R50 net   \n",
       "46    amino acid biosynthesis      R51 net   \n",
       "47    amino acid biosynthesis          R52   \n",
       "48    amino acid biosynthesis          R53   \n",
       "\n",
       "                                             Equation  \\\n",
       "0                                Glucose + ATP -> G6P   \n",
       "1                                GLYC + ATP -> Glyc3P   \n",
       "2                              Glyc3P <-> DHAP + UQH2   \n",
       "3           OA + ATP -> 9*ACCOAcyt + 7*NADH + 7*FADH2   \n",
       "4                                         G6P <-> F6P   \n",
       "5                                   F6P + ATP <-> FBP   \n",
       "6                                  FBP <-> DHAP + GAP   \n",
       "7                                        DHAP <-> GAP   \n",
       "8                            GAP <-> G3P + ATP + NADH   \n",
       "9                                         G3P <-> PEP   \n",
       "10                               PEP <-> PYRcyt + ATP   \n",
       "11                          OAAcyt + ATP -> PEP + CO2   \n",
       "12                                 G6P -> PG6 + NADPH   \n",
       "13                          PG6 -> Ru5P + CO2 + NADPH   \n",
       "14                                       Ru5P <-> R5P   \n",
       "15                                       Ru5P <-> X5P   \n",
       "16                                 X5P <-> GAP + TKC2   \n",
       "17                                 E4P + TKC2 <-> F6P   \n",
       "18                                 R5P + TKC2 <-> S7P   \n",
       "19                                 GAP + TAC3 <-> F6P   \n",
       "20                                 S7P <-> E4P + TAC3   \n",
       "21                    PYRmit -> ACCOAmit + CO2 + NADH   \n",
       "22                        ACCOAmit + OAAmit -> CITmit   \n",
       "23                                  CITmit <-> ICTmit   \n",
       "24                         ICTmit -> AKG + CO2 + NADH   \n",
       "25                        ICTmit -> AKG + CO2 + NADPH   \n",
       "26  AKG -> 0.5*SUCmit + 0.5*SUCmit + CO2 + ATP + NADH   \n",
       "27                           SUCmit <-> FUMmit + UQH2   \n",
       "28  0.5*FUMmit + 0.5*FUMmit <-> 0.5*MALmit + 0.5*M...   \n",
       "29                           MALmit <-> OAAmit + NADH   \n",
       "30                     MALmit <-> PYRmit + CO2 + NADH   \n",
       "31                       PYRcyt + CO2 + ATP -> OAAcyt   \n",
       "32                                    AA -> AC + NADH   \n",
       "33                                   AA -> AC + NADPH   \n",
       "34                              AC + ATP <-> ACCOAcyt   \n",
       "35                                  CITcyt <-> ICTcyt   \n",
       "36                       ICTcyt -> GLYOXYcyt + SUCcyt   \n",
       "37                     ACCOAcyt + GLYOXYcyt -> MALcyt   \n",
       "38                            MALcyt -> OAAcyt + NADH   \n",
       "39                                   PYRcyt -> PYRmit   \n",
       "40                              ACCOAcyt <-> ACCOAmit   \n",
       "41                                   CITmit -> CITcyt   \n",
       "42                                   SUCcyt -> SUCmit   \n",
       "43                                   MALcyt -> MALmit   \n",
       "44                                  G3P -> SER + NADH   \n",
       "45                                 SER <-> GLY + MTHF   \n",
       "46                          CO2 + MTHF + NADH <-> GLY   \n",
       "47                                 AKG + NADPH -> GLU   \n",
       "48                                   GLU + ATP -> GLN   \n",
       "\n",
       "                                reaction_ids              pathway  \\\n",
       "0                           reverse_EX_glc_e     substrate_uptake   \n",
       "1                              reverse_GLYCt                  emp   \n",
       "2                                        NaN                  emp   \n",
       "3                                    OCDCEAt     substrate_uptake   \n",
       "4                                        PGI                  emp   \n",
       "5                         PFK or reverse_FBP                  emp   \n",
       "6                                        FBA                  emp   \n",
       "7                                        TPI                  emp   \n",
       "8                                       GAPD                  emp   \n",
       "9                                        ENO                  emp   \n",
       "10                                       PYK                  emp   \n",
       "11                                      PPCK          anaplerotic   \n",
       "12                                    G6PDH2    pentose_phosphate   \n",
       "13                                       GND    pentose_phosphate   \n",
       "14                               reverse_RPI    pentose_phosphate   \n",
       "15                                       RPE    pentose_phosphate   \n",
       "16                              TKT1 or TKT2    pentose_phosphate   \n",
       "17                                      TKT2    pentose_phosphate   \n",
       "18                                      TKT1    pentose_phosphate   \n",
       "19                                      TALA    pentose_phosphate   \n",
       "20                                      TALA    pentose_phosphate   \n",
       "21                                      PDHm            tca_cycle   \n",
       "22                                       CSm            tca_cycle   \n",
       "23                                    ACONTm            tca_cycle   \n",
       "24                                    ICDHxm            tca_cycle   \n",
       "25                                    ICDHym            tca_cycle   \n",
       "26                         AKGDam and AKGDbm            tca_cycle   \n",
       "27                      SUCD2_u6m and SUCD1m            tca_cycle   \n",
       "28                                      FUMm            tca_cycle   \n",
       "29                                      MDHm            tca_cycle   \n",
       "30                                      ME1m          anaplerotic   \n",
       "31                                        PC          anaplerotic   \n",
       "32                                    ALDD2y  acetate_metabolism    \n",
       "33                                       NaN  acetate_metabolism    \n",
       "34                      ACS or reverse_ACOAH  acetate_metabolism    \n",
       "35                                       NaN                  NaN   \n",
       "36                                       ICL     glyoxylate_shunt   \n",
       "37                                     MALSp     glyoxylate_shunt   \n",
       "38                                       MDH            tca_cycle   \n",
       "39                                    PYRt2m            transport   \n",
       "40                                       NaN                  NaN   \n",
       "41                CITtam or CITtbm or CITtcm            transport   \n",
       "42                                    SUCCtm            transport   \n",
       "43  MALtm or reverse_CITtam or reverse_DICtm            transport   \n",
       "44                                       NaN                  NaN   \n",
       "45                                       NaN                  NaN   \n",
       "46                                       NaN                  NaN   \n",
       "47                                       NaN                  NaN   \n",
       "48                                       NaN                  NaN   \n",
       "\n",
       "     compartment  glucose_flux  glucose_std_err  glucose_LB  glucose_UB  ...  \\\n",
       "0        cytosol  1.000000e+02        5590.0000    100.0000    100.0000  ...   \n",
       "1        cytosol           NaN              NaN         NaN         NaN  ...   \n",
       "2        cytosol           NaN              NaN         NaN         NaN  ...   \n",
       "3        cytosol           NaN              NaN         NaN         NaN  ...   \n",
       "4        cytosol  1.615850e+01         903.3512      7.6380     25.0760  ...   \n",
       "5        cytosol  5.470010e+01        3057.7000     50.9378     57.0545  ...   \n",
       "6        cytosol  5.470010e+01        3057.7000     50.9378     57.0545  ...   \n",
       "7        cytosol  5.295380e+01        2960.1000     49.0946     55.3190  ...   \n",
       "8        cytosol  1.262018e+02        7054.7000    120.5904    130.0255  ...   \n",
       "9        cytosol  1.214871e+02        6791.1000    115.6079    125.6591  ...   \n",
       "10       cytosol  1.172436e+02        6553.8000    111.0774    126.5098  ...   \n",
       "11       cytosol  9.999800e-08           7.9151      0.0000      9.6176  ...   \n",
       "12       cytosol  6.092920e+01        3406.0000     50.2982     69.3906  ...   \n",
       "13       cytosol  6.092920e+01        3406.0000     50.2982     69.3906  ...   \n",
       "14       cytosol  2.238770e+01        1251.5000     19.0080     25.2153  ...   \n",
       "15       cytosol  3.854150e+01        2154.5000     31.2322     42.9235  ...   \n",
       "16       cytosol  3.854150e+01        2154.5000     31.2322     42.9235  ...   \n",
       "17       cytosol  1.820990e+01        1017.9000     14.4968     21.0188  ...   \n",
       "18       cytosol  2.033160e+01        1136.6000     16.7898     23.1501  ...   \n",
       "19       cytosol  2.033160e+01        1136.6000     16.7898     23.1501  ...   \n",
       "20       cytosol  2.033160e+01        1136.6000     16.7898     23.1501  ...   \n",
       "21  mitochondria  9.279170e+01        5187.0000     86.6629    100.6986  ...   \n",
       "22  mitochondria  4.028740e+01        2252.0000     28.1700     51.2108  ...   \n",
       "23  mitochondria  2.796850e+01        1563.6000     12.6008     40.3555  ...   \n",
       "24  mitochondria  2.796850e+01       89320.0000      0.0000     40.3559  ...   \n",
       "25  mitochondria  1.000200e-07       89302.0000      0.0000     30.8352  ...   \n",
       "26  mitochondria  1.596080e+01         892.6194      0.0000     29.3331  ...   \n",
       "27  mitochondria  2.827970e+01        1580.9000     15.0997     40.2079  ...   \n",
       "28  mitochondria  3.006350e+01        1680.6000     17.1305     41.8427  ...   \n",
       "29  mitochondria  4.028740e+01        2252.0000     28.1700     51.2108  ...   \n",
       "30  mitochondria -1.022390e+01         571.4598    -11.2237     -4.7120  ...   \n",
       "31       cytosol  1.328400e+00          74.2421      0.0000      5.7862  ...   \n",
       "32       cytosol  1.573800e-05       60003.0000      0.0000      1.1631  ...   \n",
       "33       cytosol  9.542000e-01       60003.0000      0.0000      2.1173  ...   \n",
       "34       cytosol  2.343700e+00         131.0054      1.3221      3.5411  ...   \n",
       "35           NaN  1.231890e+01         688.7175      9.9873     20.8181  ...   \n",
       "36       cytosol  1.231890e+01         688.7175      9.9873     20.8181  ...   \n",
       "37       cytosol  1.231890e+01         688.7175      9.9873     20.8181  ...   \n",
       "38       cytosol  1.231890e+01         688.5462      9.9871     21.7042  ...   \n",
       "39  mitochondria  1.162532e+02        6498.4000    109.8734    123.6819  ...   \n",
       "40           NaN -5.025110e+01        2809.2000    -61.8069    -45.1953  ...   \n",
       "41  mitochondria  1.231890e+01         688.7175      9.9873     20.8181  ...   \n",
       "42  mitochondria  1.231890e+01         688.7175      9.9873     20.8181  ...   \n",
       "43  mitochondria  1.000100e-07           9.2808      0.0000      6.1103  ...   \n",
       "44           NaN  4.714600e+00         263.5628      4.0957      5.4976  ...   \n",
       "45           NaN  1.973200e+00         110.3147      1.4156      2.5653  ...   \n",
       "46           NaN  3.585000e-01          20.0475     -0.2078      0.8764  ...   \n",
       "47           NaN  4.058240e+01        2268.6000     36.9632     45.1679  ...   \n",
       "48           NaN  6.205700e+00         346.9084      5.6858      6.8391  ...   \n",
       "\n",
       "    glycerol_flux  glycerol_std_err  glycerol_LB  glycerol_UB  glycerol_ΔB  \\\n",
       "0             NaN               NaN          NaN          NaN          NaN   \n",
       "1      100.000000      5.974500e-11     100.0000     100.0000       0.0000   \n",
       "2      100.000000      5.974500e-11     100.0000     100.0000       0.0000   \n",
       "3             NaN               NaN          NaN          NaN          NaN   \n",
       "4      -19.168800      1.740200e+00     -26.9472     -15.8469      11.1003   \n",
       "5      -16.913700      6.263000e-01     -19.3382     -15.6909       3.6473   \n",
       "6      -16.913700      6.263000e-01     -19.3382     -15.6909       3.6473   \n",
       "7       82.027300      6.319000e-01      78.9156      83.2640       4.3484   \n",
       "8       65.802800      7.356000e-01      63.6523      67.2932       3.6409   \n",
       "9       63.201900      7.767000e-01      61.1082      64.7828       3.6746   \n",
       "10      60.628300      7.959000e-01      58.5800      62.2551       3.6751   \n",
       "11       0.000002      2.580800e-11       0.0000       0.7782       0.7782   \n",
       "12       5.273000      1.697800e+00       2.0588      15.8682      13.8094   \n",
       "13       5.273000      1.697800e+00       2.0588      15.8682      13.8094   \n",
       "14       3.017900      5.694000e-01       1.9264       6.5241       4.5977   \n",
       "15       2.255100      1.128700e+00       0.1059       7.6108       7.5049   \n",
       "16       2.255100      1.128700e+00       0.1059       7.6108       7.5049   \n",
       "17       0.484200      5.628000e-01      -0.5867       4.0518       4.6385   \n",
       "18       1.771000      5.660000e-01       0.6921       4.4382       3.7461   \n",
       "19       1.771000      5.660000e-01       0.6921       4.4382       3.7461   \n",
       "20       1.771000      5.660000e-01       0.6921       4.4382       3.7461   \n",
       "21      43.592200      9.147000e-01      41.5129      45.4798       3.9669   \n",
       "22      13.956100      1.206500e+00      12.5923      16.5250       3.9327   \n",
       "23       8.174700      1.311100e+00       7.1646      10.9497       3.7851   \n",
       "24       4.161800      3.533800e+05       0.0000      10.9497      10.9497   \n",
       "25       4.012900      3.533800e+05       0.0000       9.1309       9.1309   \n",
       "26       0.892300      1.383800e+00       0.0000       3.8277       3.8277   \n",
       "27       6.673700      1.277000e+00       5.4743       9.4004       3.9261   \n",
       "28       7.755500      1.266400e+00       6.5658      10.4589       3.8931   \n",
       "29      13.956100      1.206500e+00      12.5923      16.5250       3.9327   \n",
       "30      -6.200600      6.770000e-02      -6.2882      -5.9066       0.3816   \n",
       "31       3.012200      2.388000e-01       2.5256       3.5355       1.0099   \n",
       "32       0.000002      2.441600e+05       0.0000       0.2015       0.2015   \n",
       "33       1.095600      2.441600e+05       0.0000       1.3082       1.3082   \n",
       "34       1.938200      9.150000e-02       1.6632       2.1408       0.4776   \n",
       "35       5.781400      2.544000e-01       5.3203       6.3999       1.0796   \n",
       "36       5.781400      2.544000e-01       5.3203       6.3999       1.0796   \n",
       "37       5.781400      2.544000e-01       5.3203       6.3999       1.0796   \n",
       "38       5.781400      2.544000e-01       5.3203       6.3997       1.0794   \n",
       "39      57.821000      8.105000e-01      55.5389      59.4518       3.9129   \n",
       "40     -28.269600      4.550000e-01     -29.0546     -27.2949       1.7597   \n",
       "41       5.781400      2.544000e-01       5.3203       6.3999       1.0796   \n",
       "42       5.781400      2.544000e-01       5.3203       6.3999       1.0796   \n",
       "43       0.000002      3.522900e-11       0.0000       0.3057       0.3057   \n",
       "44       2.600900      5.770000e-02       2.4696       2.7419       0.2723   \n",
       "45       0.938300      4.880000e-02       0.8274       1.0741       0.2467   \n",
       "46      -0.041000      4.610000e-02      -0.1422       0.0949       0.2371   \n",
       "47      25.129200      2.759000e-01      24.5244      25.5919       1.0675   \n",
       "48       3.763600      4.110000e-02       3.6714       3.8168       0.1454   \n",
       "\n",
       "    oleic_acid_flux  oleic_acid_std_err  oleic_acid_LB  oleic_acid_UB  \\\n",
       "0               NaN                 NaN            NaN            NaN   \n",
       "1               NaN                 NaN            NaN            NaN   \n",
       "2               NaN                 NaN            NaN            NaN   \n",
       "3      1.000000e+02        2.594100e-10       100.0000       100.0000   \n",
       "4     -1.035480e+02        3.191300e+00      -345.1830       -82.7140   \n",
       "5     -9.801660e+01        3.020800e+00      -167.8457       -86.4527   \n",
       "6     -9.801660e+01        3.020800e+00      -167.8457       -86.4527   \n",
       "7     -1.083464e+02        3.339200e+00      -176.4324       -95.9590   \n",
       "8     -2.051634e+02        6.323000e+00      -256.9114      -184.8529   \n",
       "9     -2.169470e+02        7.763400e+00      -262.8657      -194.3019   \n",
       "10     1.412210e+01        1.434390e+01         0.0000       112.2861   \n",
       "11     2.421320e+02        1.836350e+01       204.4483       372.5395   \n",
       "12     1.736070e+01        5.350000e-01         0.0000       273.5399   \n",
       "13     1.736070e+01        5.350000e-01         0.0000       273.5399   \n",
       "14     1.182930e+01        3.646000e-01         5.8507        96.2027   \n",
       "15     5.531400e+00        1.705000e-01        -6.2991       177.3372   \n",
       "16     5.531400e+00        1.705000e-01        -6.2991       177.3372   \n",
       "17    -7.205800e-07        3.578900e-09        -6.0327        86.3696   \n",
       "18     5.531400e+00        1.705000e-01        -0.2663        90.9676   \n",
       "19     5.531400e+00        1.705000e-01        -0.2663        90.9676   \n",
       "20     5.531400e+00        1.705000e-01        -0.2663        90.9676   \n",
       "21     8.046600e-07        6.512000e-11         0.0000        96.2850   \n",
       "22     3.644949e+02        1.751540e+01       347.2815       430.2887   \n",
       "23     5.138920e+01        2.615360e+01        21.0202       110.6330   \n",
       "24     2.058700e-06        3.195200e+05         0.0000        85.2036   \n",
       "25     5.138920e+01        3.195200e+05         0.0000       110.6141   \n",
       "26     2.509820e+01        2.696390e+01         0.0000        86.2498   \n",
       "27     3.382039e+02        1.832330e+01       319.8737       410.6453   \n",
       "28     3.418693e+02        1.821070e+01       323.6948       413.3904   \n",
       "29     3.644949e+02        1.751540e+01       347.2815       430.2887   \n",
       "30     2.732400e+01        8.421000e-01       -23.5867       120.2857   \n",
       "31     1.532170e+01        1.435670e+01         0.0000        62.3378   \n",
       "32     9.578000e-01        3.289600e+00         0.0000        24.7619   \n",
       "33     1.001600e-07        9.546500e-09         0.0000        23.8039   \n",
       "34     5.889500e+00        3.204700e+00         3.8578        29.7333   \n",
       "35     3.131057e+02        8.793700e+00       291.4722       404.4560   \n",
       "36     3.131057e+02        8.793700e+00       291.4722       404.4560   \n",
       "37     3.131057e+02        8.793700e+00       291.4722       404.4560   \n",
       "38     2.631561e+02        7.276300e+00       239.1927       390.4334   \n",
       "39     8.492400e-07        2.092900e-10         0.0000       118.6184   \n",
       "40     3.691600e+02        1.737200e+01       316.6482       423.2757   \n",
       "41     3.131057e+02        8.793700e+00       291.4722       404.4560   \n",
       "42     3.131057e+02        8.793700e+00       291.4722       404.4560   \n",
       "43     4.994960e+01        1.539400e+00         0.0000       141.0555   \n",
       "44     1.178360e+01        1.887400e+00         0.0000        12.7834   \n",
       "45     5.719000e+00        1.761500e+00        -6.0935         6.4612   \n",
       "46     9.873000e-01        1.671200e+00       -10.8917         1.5284   \n",
       "47     9.096050e+01        2.834100e+00        69.6159       115.4479   \n",
       "48     1.936010e+01        5.967000e-01        14.6939        20.1824   \n",
       "\n",
       "    oleic_acid_ΔB  \n",
       "0             NaN  \n",
       "1             NaN  \n",
       "2             NaN  \n",
       "3          0.0000  \n",
       "4        262.4690  \n",
       "5         81.3930  \n",
       "6         81.3930  \n",
       "7         80.4734  \n",
       "8         72.0585  \n",
       "9         68.5638  \n",
       "10       112.2861  \n",
       "11       168.0912  \n",
       "12       273.5399  \n",
       "13       273.5399  \n",
       "14        90.3520  \n",
       "15       183.6363  \n",
       "16       183.6363  \n",
       "17        92.4023  \n",
       "18        91.2339  \n",
       "19        91.2339  \n",
       "20        91.2339  \n",
       "21        96.2850  \n",
       "22        83.0072  \n",
       "23        89.6128  \n",
       "24        85.2036  \n",
       "25       110.6141  \n",
       "26        86.2498  \n",
       "27        90.7716  \n",
       "28        89.6956  \n",
       "29        83.0072  \n",
       "30       143.8724  \n",
       "31        62.3378  \n",
       "32        24.7619  \n",
       "33        23.8039  \n",
       "34        25.8755  \n",
       "35       112.9838  \n",
       "36       112.9838  \n",
       "37       112.9838  \n",
       "38       151.2407  \n",
       "39       118.6184  \n",
       "40       106.6275  \n",
       "41       112.9838  \n",
       "42       112.9838  \n",
       "43       141.0555  \n",
       "44        12.7834  \n",
       "45        12.5547  \n",
       "46        12.4201  \n",
       "47        45.8320  \n",
       "48         5.4885  \n",
       "\n",
       "[49 rows x 21 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "central_rxn_df = pd.read_excel('../data/13c_mfa/INCA_model_10232023_GR.xlsx', sheet_name='flux_data')\n",
    "\n",
    "# calculate the number of reactions in the 13C MFA that are mapped to the GSM\n",
    "mapped_rxn_df = central_rxn_df.dropna(subset = [\"reaction_ids\"])\n",
    "\n",
    "print(f'There are {len(mapped_rxn_df)} reactions in the 13C MFA that are mapped to the GSM')\n",
    "\n",
    "central_rxn_df.head(49)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Develop Flux Map Generator Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: './images/unlabeled_flux_map.png'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb Cell 11\u001b[0m line \u001b[0;36m6\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=61'>62</a>\u001b[0m     plt\u001b[39m.\u001b[39msavefig(file_name)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=62'>63</a>\u001b[0m     plt\u001b[39m.\u001b[39mshow()\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=64'>65</a>\u001b[0m generate_flux_map(\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=65'>66</a>\u001b[0m     flux_df\u001b[39m=\u001b[39;49m\u001b[39mNone\u001b[39;49;00m, \n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=66'>67</a>\u001b[0m     glucose_column_name\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mglucose_flux\u001b[39;49m\u001b[39m'\u001b[39;49m, \n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=67'>68</a>\u001b[0m     glycerol_column_name\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39mglycerol_flux\u001b[39;49m\u001b[39m'\u001b[39;49m,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=68'>69</a>\u001b[0m     oleic_acid_column_name\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39moleic_acid_flux\u001b[39;49m\u001b[39m'\u001b[39;49m,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=69'>70</a>\u001b[0m     file_name\u001b[39m=\u001b[39;49m\u001b[39m'\u001b[39;49m\u001b[39m../figures/test_flux_map.png\u001b[39;49m\u001b[39m'\u001b[39;49m,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=70'>71</a>\u001b[0m )\n",
      "\u001b[1;32m/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb Cell 11\u001b[0m line \u001b[0;36m1\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m fig, ax \u001b[39m=\u001b[39m plt\u001b[39m.\u001b[39msubplots(figsize\u001b[39m=\u001b[39m(\u001b[39m15\u001b[39m, \u001b[39m20\u001b[39m), dpi\u001b[39m=\u001b[39m\u001b[39m50\u001b[39m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m \u001b[39m# load unlabeled image and set as plot background\u001b[39;00m\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=15'>16</a>\u001b[0m unlabeled_image \u001b[39m=\u001b[39m plt\u001b[39m.\u001b[39;49mimread(\u001b[39m'\u001b[39;49m\u001b[39m./images/unlabeled_flux_map.png\u001b[39;49m\u001b[39m'\u001b[39;49m)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=16'>17</a>\u001b[0m imagebox \u001b[39m=\u001b[39m OffsetImage(unlabeled_image)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/garrettroell/yarrowia_eflux2/notebooks/C_13C_pFBA.ipynb#X13sZmlsZQ%3D%3D?line=17'>18</a>\u001b[0m imagebox\u001b[39m.\u001b[39mimage\u001b[39m.\u001b[39maxes \u001b[39m=\u001b[39m ax\n",
      "File \u001b[0;32m~/yarrowia_eflux2/venv/lib/python3.10/site-packages/matplotlib/pyplot.py:2195\u001b[0m, in \u001b[0;36mimread\u001b[0;34m(fname, format)\u001b[0m\n\u001b[1;32m   2193\u001b[0m \u001b[39m@_copy_docstring_and_deprecators\u001b[39m(matplotlib\u001b[39m.\u001b[39mimage\u001b[39m.\u001b[39mimread)\n\u001b[1;32m   2194\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mimread\u001b[39m(fname, \u001b[39mformat\u001b[39m\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[0;32m-> 2195\u001b[0m     \u001b[39mreturn\u001b[39;00m matplotlib\u001b[39m.\u001b[39;49mimage\u001b[39m.\u001b[39;49mimread(fname, \u001b[39mformat\u001b[39;49m)\n",
      "File \u001b[0;32m~/yarrowia_eflux2/venv/lib/python3.10/site-packages/matplotlib/image.py:1563\u001b[0m, in \u001b[0;36mimread\u001b[0;34m(fname, format)\u001b[0m\n\u001b[1;32m   1556\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(fname, \u001b[39mstr\u001b[39m) \u001b[39mand\u001b[39;00m \u001b[39mlen\u001b[39m(parse\u001b[39m.\u001b[39murlparse(fname)\u001b[39m.\u001b[39mscheme) \u001b[39m>\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[1;32m   1557\u001b[0m     \u001b[39m# Pillow doesn't handle URLs directly.\u001b[39;00m\n\u001b[1;32m   1558\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m   1559\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mPlease open the URL for reading and pass the \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   1560\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mresult to Pillow, e.g. with \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   1561\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39m``np.array(PIL.Image.open(urllib.request.urlopen(url)))``.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   1562\u001b[0m         )\n\u001b[0;32m-> 1563\u001b[0m \u001b[39mwith\u001b[39;00m img_open(fname) \u001b[39mas\u001b[39;00m image:\n\u001b[1;32m   1564\u001b[0m     \u001b[39mreturn\u001b[39;00m (_pil_png_to_float_array(image)\n\u001b[1;32m   1565\u001b[0m             \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(image, PIL\u001b[39m.\u001b[39mPngImagePlugin\u001b[39m.\u001b[39mPngImageFile) \u001b[39melse\u001b[39;00m\n\u001b[1;32m   1566\u001b[0m             pil_to_array(image))\n",
      "File \u001b[0;32m~/yarrowia_eflux2/venv/lib/python3.10/site-packages/PIL/ImageFile.py:105\u001b[0m, in \u001b[0;36mImageFile.__init__\u001b[0;34m(self, fp, filename)\u001b[0m\n\u001b[1;32m    101\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdecodermaxblock \u001b[39m=\u001b[39m MAXBLOCK\n\u001b[1;32m    103\u001b[0m \u001b[39mif\u001b[39;00m is_path(fp):\n\u001b[1;32m    104\u001b[0m     \u001b[39m# filename\u001b[39;00m\n\u001b[0;32m--> 105\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfp \u001b[39m=\u001b[39m \u001b[39mopen\u001b[39;49m(fp, \u001b[39m\"\u001b[39;49m\u001b[39mrb\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n\u001b[1;32m    106\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mfilename \u001b[39m=\u001b[39m fp\n\u001b[1;32m    107\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_exclusive_fp \u001b[39m=\u001b[39m \u001b[39mTrue\u001b[39;00m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: './images/unlabeled_flux_map.png'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAmQAAAMZCAYAAACu2igXAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjcuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/bCgiHAAAACXBIWXMAAAexAAAHsQEGxWGGAAAWt0lEQVR4nO3cT2jf9R3H8XdLByUhqcgOBX8Ye5zx4C497KB4aKKrBElBtJSIRdP9K8Wl1BYRdpiHVTttLW0XgoYY+sfEOMhhWxish9GD1yI7B+NBD0rXrgwh/nYoRsL8znY1eW2/Ph7wOfy+f355w+fy5Ntfvxuqql0AAMRsTA8AAHCnE2QAAGGCDAAgTJABAIQ1Btm2bdtqYmKiZmZmVh3v7++v6enpmp6erv7+/jUfEADgTtD+T2tmZmbV5/Hx8faWLVvavb297TNnzvzHey3LsizLsqxvX5vqFm3ZsqWuXLlSVVU9PT2rzg0MDNTg4GBt3769Pvjgg1v9agCAjnLffffVrl27vvW6Ww6yK1euVG9vb7Xb7bp69eqqcwsLC7WwsFDHjh2rsbGxW/1qAICOcuzYsZu6rjHI7r777nrllVfqhz/8YR0+fLjuv//+GhkZqePHj9ebb75ZVVVHjx79bqYFALiDNQbZZ599Vj/96U//7fiHH35YzzzzzJoOBQBwJ/HaCwCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAEBYY5B1dXXV5ORkjY+P1+7du1eOP/bYY3XhwoW6cOFC7dixY12GBADoZI1BNjw8XLOzszU6OlpDQ0Mrxx955JH65S9/WQcOHKgf//jH6zIkAEAn29R0otVq1eXLl6uqanl5eeX4+fPna2ZmpjZs2FCjo6Or7hkYGKjBwcHq6+tbo3EBADpP4xOypaWlarVaNy7a+PVlR44cqYcffrgeeuihOnz48Kp7FhYWamxsrBYXF9doXACAztP4hGxubq5OnjxZO3furPn5+ZqamqqRkZGam5uriYmJqqr64x//uG6DAgB0qsYgu379eu3du3fl89mzZ6uq6ty5c3Xu3Lm1nwwA4A7htRcAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAsE1NJ7q6uurUqVP1xRdf1MWLF+vs2bNVVbV169Y6cuRIbdiwoc6fP1+XLl1at2EBADpR4xOy4eHhmp2drdHR0RoaGlo5fvDgwbp69Wp9+eWXtbS0tC5DAgB0ssYnZK1Wqy5fvlxVVcvLyyvH+/v768CBA/Xpp5/Wq6++Ws8///zKuYGBgRocHKy+vr41HBkAoLM0PiFbWlqqVqt146KNG1cd//zzz+vatWu1efPmVfcsLCzU2NhYLS4urtG4AACdp/EJ2dzcXJ08ebJ27txZ8/PzNTU1VSMjI/Xb3/62jh49Wu12u06fPr2eswIAdKTGILt+/Xrt3bt35fNXP+r/29/+Vs8+++zaTwYAcIfw2gsAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAWGOQdXV11eTkZI2Pj9fu3btXnXvggQfqk08+qe7u7jUfEACg0zUG2fDwcM3Oztbo6GgNDQ2tHN+0aVM999xz9Yc//GFdBgQA6HSNQdZqteqjjz6qqqrl5eWV4wcPHqwTJ05Uu93+t3sGBgbq2LFj1dfXtwajAgB0psYgW1paqlardeOijV9f9uCDD9b+/ftr+/bttW/fvlX3LCws1NjYWC0uLq7RuAAAnacxyObm5mrXrl116tSpmp+fr6mpqaqqeuqpp+qFF16oDz74oH73u9+t26AAAJ1qU9OJ69ev1969e1c+nz17dtX5Z599du2mAgC4g3jtBQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBsU9OJrq6uOnXqVH3xxRd18eLFOnv2bFVVvfjii7Vt27b6/ve/XwcOHKiPP/543YYFAOhEjU/IhoeHa3Z2tkZHR2toaGjl+G9+85v6yU9+UtPT0/XII4+sy5AAAJ2s8QlZq9Wqy5cvV1XV8vLyqnPd3d315JNP1ujo6KrjAwMDNTg4WH19fWswKgBAZ2p8Qra0tFStVuvGRRu/vqynp6dOnz5dhw4dqmvXrq26Z2FhocbGxmpxcXGNxgUA6DyNQTY3N1e7du2qU6dO1fz8fE1NTVVV1eTkZN1111310ksv+SdLAIDvQOM/WV6/fr327t278vmrH/Xv2rVr7acCALiDeO0FAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIEyQAQCECTIAgDBBBgAQJsgAAMIEGQBAmCADAAgTZAAAYYIMACBMkAEAhAkyAIAwQQYAECbIAADCBBkAQJggAwAIE2QAAGGCDAAgTJABAIQJMgCAMEEGABAmyAAAwgQZAECYIAMACBNkAABhggwAIKwxyLq6umpycrLGx8dr9+7dK8f7+/trenq6pqenq7+/f12GBADoZI1BNjw8XLOzszU6OlpDQ0Mrxw8cOFA///nP62c/+1nt379/XYYEAOhkm5pOtFqtunz5clVVLS8vrxzfsmVLXblypaqqenp6Vt0zMDBQg4ODNTg4uBazEtLX11eLi4vpMfgO2dPOYj87i/3sLNu3b7/pa9vftPbs2dPeuXNnu6ra586dWzk+Pj7e7u3tbff09LTPnDnzjfceO3bsG49b/5/LfnbesqedtexnZy372VnrZvez8QnZ3NxcnTx5snbu3Fnz8/M1NTVVIyMjdfz48XrzzTerquro0aPfeO+f/vSnpq/l/5D97Dz2tLPYz85iPzvLze7nhrpRZgAAhHjtBQBA2G0HmddjdJam/XzxxRfrzJkzNTs7W/fcc09wQm5V055WVT3wwAP1ySefVHd3d2g6blXTfm7durWOHz9eJ06cqB/96EfBCbkVTfv52GOP1YULF+rChQu1Y8eO4ITcqm3bttXExETNzMysOn4zXXRbP1bbs2dP+/HHH29XVfv8+fMrx8fHx9tbtmxp9/b2Nv743/rfW037+dV64okn2nv27InPad3+nm7atKn9xhtvtCcnJ9vd3d3xOa3b28/XXnut/etf/7r9xhtvtO+99974nNbt7efRo0fb99xzT3vr1q3t119/PT6ndetrZmZm1edv66LbfkLWarXqo48+qqpvfj3G3//+9397PQb/u5r2s6qqu7u7nnzyyfr9738fmIz/VtOeHjx4sE6cOFHtdjs1Gv+Fpv3s7++vqamp+tWvflUvv/xyajxuUdN+nj9/vmZmZur999+vt956KzUe36Fv66LbDrKlpaVqtVo3vmzj11935cqV6u3trZ6enrp69ert/hnWSdN+9vT01OnTp+vQoUN17dq11Hj8F5r29MEHH6z9+/fX9u3ba9++fanxuEVN+7m0tFSff/55Xbt2rTZv3pwaj1vUtJ9Hjhyphx9+uB566KE6fPhwajy+Q9/WRbf9vyy7urrq5MmT9c9//rP++te/1qOPPlojIyPV399fhw4dqqobr8f48MMPb+fPsE6a9vO9996r733ve/Xxxx/Xu+++W3/5y1/So3KTmvb0K2+//Xb94he/qH/84x/BKblZTfv5gx/8oA4dOlTtdrsmJibq0qVL6VG5CU37+fTTT9ejjz5aVVV//vOf65133glPys26++6765VXXqkdO3bUxMRE3X///TfVRV57AQAQ5rUXAABhggwAIEyQAQCECTIAgDBBBgAQ9i99sCVbeJqddQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 750x1000 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.offsetbox import (OffsetImage, AnnotationBbox, TextArea)\n",
    "\n",
    "# This function takes in a flux data frame and column name, and generates a flux map using the flux values in the specified column\n",
    "def generate_flux_map(\n",
    "    flux_df=None, \n",
    "    glucose_column_name='', \n",
    "    glycerol_column_name='', \n",
    "    oleic_acid_column_name='', \n",
    "    file_name='',\n",
    "):\n",
    "    # define plot area\n",
    "    fig, ax = plt.subplots(figsize=(15, 20), dpi=50)\n",
    "    \n",
    "    # load unlabeled image and set as plot background\n",
    "    unlabeled_image = plt.imread('./images/unlabeled_flux_map.png')\n",
    "    imagebox = OffsetImage(unlabeled_image)\n",
    "    imagebox.image.axes = ax\n",
    "    xy = (0.5, 0.5)\n",
    "    ab = AnnotationBbox(imagebox, xy, frameon=False)\n",
    "    ax.add_artist(ab)\n",
    "\n",
    "    # loop over each reaction in the dataframe\n",
    "    for _, row in flux_df.iterrows():\n",
    "        # check that there is a location for the reaction's flux\n",
    "        if not pd.isnull(row['Location on map']):\n",
    "            # get the flux value\n",
    "            flux_value = row[column_name]\n",
    "\n",
    "            # get the location of the reaction's flux on the map as a tuple\n",
    "            location =  row['Location on map'].replace('(', '').replace(')', '')\n",
    "            location_list = location.split(',')\n",
    "            location_tuple = tuple((int(location_list[0]), int(location_list[1])))\n",
    "\n",
    "            # create a text area with the flux value\n",
    "            offsetbox = TextArea(\n",
    "                f'{flux_value:.1f}', \n",
    "                textprops=dict(fontsize=22, color='black')\n",
    "            )\n",
    "            \n",
    "            # creates an annotation box, which is a box that can be placed on the plot area.\n",
    "            ab = AnnotationBbox(\n",
    "                offsetbox, \n",
    "                xy,\n",
    "                xybox=location_tuple,\n",
    "                xycoords='data',\n",
    "                boxcoords=\"offset points\",\n",
    "                frameon=False\n",
    "            )\n",
    "\n",
    "            # add the annotation box to the plot area\n",
    "            ax.add_artist(ab)\n",
    "\n",
    "    # ensure that the axes have minimal styles \n",
    "    ax.set_xlim(0, 1)\n",
    "    ax.set_ylim(0, 1)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_yticks([])\n",
    "    ax.axis(\"off\")\n",
    "    \n",
    "    # save and show the plot\n",
    "    plt.savefig(file_name)\n",
    "    plt.show()\n",
    "\n",
    "generate_flux_map(\n",
    "    flux_df=None, \n",
    "    glucose_column_name='glucose_flux', \n",
    "    glycerol_column_name='glycerol_flux',\n",
    "    oleic_acid_column_name='oleic_acid_flux',\n",
    "    file_name='../figures/test_flux_map.png',\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Glucose"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define a function to build constraint strings from 13C-MFA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GAPD >= 8    120.5904\\nName: glucose_LB, dtype: float64, GAPD <= 8    130.0255\\nName: glucose_UB, dtype: float64, CSm >= 22    28.17\\nName: glucose_LB, dtype: float64, CSm <= 22    51.2108\\nName: glucose_UB, dtype: float64, GND >= 13    50.2982\\nName: glucose_LB, dtype: float64, GND <= 13    69.3906\\nName: glucose_UB, dtype: float64'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def build_13c_constraint_string(central_rxn_df, lower_bound_row, upper_bound_row):\n",
    "    \n",
    "    # key node reactions to constrain\n",
    "    key_pathway_reactions = {\n",
    "        'EMP': 'GAPD',\n",
    "        'TCA': 'CSm',\n",
    "        'PPP': 'GND',\n",
    "    }\n",
    "    \n",
    "    constraint_strings = []\n",
    "\n",
    "    for pathway, reaction_ids in key_pathway_reactions.items():\n",
    "        row = central_rxn_df.loc[central_rxn_df['reaction_ids'] == reaction_ids]\n",
    "        # get the lower bound for this reaction\n",
    "        lower_bound = row[lower_bound_row]\n",
    "        upper_bound = row[upper_bound_row]\n",
    "\n",
    "        constraint_string = make_rxn_constraint_string(reaction_ids, lower_bound, upper_bound)\n",
    "        # print(constraint_string)\n",
    "        # print(pathway, reaction)\n",
    "        constraint_strings.append(constraint_string)\n",
    "\n",
    "    full_constraint_string = ', '.join(constraint_strings)\n",
    "    return full_constraint_string\n",
    "\n",
    "    \n",
    "\n",
    "build_13c_constraint_string(central_rxn_df, 'glucose_LB', 'glucose_UB')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Build glucose constraint string from 13C-MFA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100 EX_glc_e\n",
      "10000 EX_h2o_e\n",
      "10000 EX_h_e\n",
      "10000 EX_nh4_e\n",
      "10000 EX_o2_e\n",
      "10000 EX_pi_e\n",
      "10000 EX_so4_e\n",
      "13C-MFA bounds: 100.0 - 100.0\n",
      "EX_glc_e >= -100.0, EX_glc_e <= -100.0\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 7.638 - 25.076\n",
      "PGI >= 7.638, PGI <= 25.076\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 50.9378 - 57.0545\n",
      "PFK - FBP >= 50.9378, PFK - FBP <= 57.0545\n",
      "max biomass flux: 10.01608296711\n",
      "\n",
      "13C-MFA bounds: 50.9378 - 57.0545\n",
      "FBA >= 50.9378, FBA <= 57.0545\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 49.0946 - 55.319\n",
      "TPI >= 49.0946, TPI <= 55.319\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 120.5904 - 130.0255\n",
      "GAPD >= 120.5904, GAPD <= 130.0255\n",
      "max biomass flux: 10.01608296711\n",
      "\n",
      "13C-MFA bounds: 115.6079 - 125.6591\n",
      "ENO >= 115.6079, ENO <= 125.6591\n",
      "max biomass flux: 10.016082967111\n",
      "\n",
      "13C-MFA bounds: 111.0774 - 126.5098\n",
      "PYK >= 111.0774, PYK <= 126.5098\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 0.0 - 9.6176\n",
      "PPCK >= 0.0, PPCK <= 9.6176\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 50.2982 - 69.3906\n",
      "G6PDH2 >= 50.2982, G6PDH2 <= 69.3906\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 50.2982 - 69.3906\n",
      "GND >= 50.2982, GND <= 69.3906\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 19.008 - 25.2153\n",
      "RPI >= -25.2153, RPI <= -19.008\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 31.2322 - 42.9235\n",
      "RPE >= 31.2322, RPE <= 42.9235\n",
      "max biomass flux: 10.008547008754\n",
      "\n",
      "13C-MFA bounds: 31.2322 - 42.9235\n",
      "TKT1 + TKT2 >= 31.2322, TKT1 + TKT2 <= 42.9235\n",
      "max biomass flux: 10.008547008754\n",
      "\n",
      "13C-MFA bounds: 14.4968 - 21.0188\n",
      "TKT2 >= 14.4968, TKT2 <= 21.0188\n",
      "max biomass flux: 10.011593621766\n",
      "\n",
      "13C-MFA bounds: 16.7898 - 23.1501\n",
      "TKT1 >= 16.7898, TKT1 <= 23.1501\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 16.7898 - 23.1501\n",
      "TALA >= 16.7898, TALA <= 23.1501\n",
      "max biomass flux: 10.016082967111\n",
      "\n",
      "13C-MFA bounds: 16.7898 - 23.1501\n",
      "TALA >= 16.7898, TALA <= 23.1501\n",
      "max biomass flux: 10.016082967111\n",
      "\n",
      "13C-MFA bounds: 86.6629 - 100.6986\n",
      "PDHm >= 86.6629, PDHm <= 100.6986\n",
      "max biomass flux: 9.802659712939\n",
      "\n",
      "13C-MFA bounds: 28.17 - 51.2108\n",
      "CSm >= 28.17, CSm <= 51.2108\n",
      "max biomass flux: 10.003149048171\n",
      "\n",
      "13C-MFA bounds: 12.6008 - 40.3555\n",
      "ACONTm >= 12.6008, ACONTm <= 40.3555\n",
      "max biomass flux: 10.016082967104\n",
      "\n",
      "13C-MFA bounds: 0.0 - 40.3559\n",
      "ICDHxm >= 0.0, ICDHxm <= 40.3559\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 0.0 - 30.8352\n",
      "ICDHym >= 0.0, ICDHym <= 30.8352\n",
      "max biomass flux: 10.016082967126\n",
      "\n",
      "13C-MFA bounds: 0.0 - 29.3331\n",
      "AKGDam >= 0.0, AKGDam <= 29.3331, AKGDbm >= 0.0, AKGDbm <= 29.3331\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 15.0997 - 40.2079\n",
      "SUCD2_u6m >= 15.0997, SUCD2_u6m <= 40.2079, SUCD1m >= 15.0997, SUCD1m <= 40.2079\n",
      "max biomass flux: 9.806192042737\n",
      "\n",
      "13C-MFA bounds: 17.1305 - 41.8427\n",
      "FUMm >= 17.1305, FUMm <= 41.8427\n",
      "max biomass flux: 0.0\n",
      "\n",
      "13C-MFA bounds: 28.17 - 51.2108\n",
      "MDHm >= 28.17, MDHm <= 51.2108\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: -11.2237 - -4.712\n",
      "ME1m >= -11.2237, ME1m <= -4.712\n",
      "max biomass flux: 0.0\n",
      "\n",
      "13C-MFA bounds: 0.0 - 5.7862\n",
      "PC >= 0.0, PC <= 5.7862\n",
      "max biomass flux: 0.0\n",
      "\n",
      "13C-MFA bounds: 0.0 - 1.1631\n",
      "ALDD2y >= 0.0, ALDD2y <= 1.1631\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 1.3221 - 3.5411\n",
      "ACS - ACOAH >= 1.3221, ACS - ACOAH <= 3.5411\n",
      "max biomass flux: 10.012969934589\n",
      "\n",
      "13C-MFA bounds: 9.9873 - 20.8181\n",
      "ICL >= 9.9873, ICL <= 20.8181\n",
      "max biomass flux: 9.98525156034\n",
      "\n",
      "13C-MFA bounds: 9.9873 - 20.8181\n",
      "MALSp >= 9.9873, MALSp <= 20.8181\n",
      "max biomass flux: 0.0\n",
      "\n",
      "13C-MFA bounds: 9.9871 - 21.7042\n",
      "MDH >= 9.9871, MDH <= 21.7042\n",
      "max biomass flux: 10.016082967112\n",
      "\n",
      "13C-MFA bounds: 8.6023 - 10.3466\n",
      "biomass_C + biomass_glucose + biomass_oil >= 8.6023, biomass_C + biomass_glucose + biomass_oil <= 10.3466\n",
      "max biomass flux: 10.016082967107\n",
      "\n",
      "13C-MFA bounds: 457.2097 - 813.9649\n",
      "ATPM >= 457.2097, ATPM <= 813.9649\n",
      "max biomass flux: 7.344359516703\n",
      "\n",
      "1391\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'EX_glc_e >= -100.0, EX_glc_e <= -100.0, PGI >= 7.638, PGI <= 25.076, PFK - FBP >= 50.9378, PFK - FBP <= 57.0545, FBA >= 50.9378, FBA <= 57.0545, TPI >= 49.0946, TPI <= 55.319, GAPD >= 120.5904, GAPD <= 130.0255, ENO >= 115.6079, ENO <= 125.6591, PYK >= 111.0774, PYK <= 126.5098, PPCK >= 0.0, PPCK <= 9.6176, G6PDH2 >= 50.2982, G6PDH2 <= 69.3906, GND >= 50.2982, GND <= 69.3906, RPI >= -25.2153, RPI <= -19.008, RPE >= 31.2322, RPE <= 42.9235, TKT1 + TKT2 >= 31.2322, TKT1 + TKT2 <= 42.9235, TKT2 >= 14.4968, TKT2 <= 21.0188, TKT1 >= 16.7898, TKT1 <= 23.1501, TALA >= 16.7898, TALA <= 23.1501, TALA >= 16.7898, TALA <= 23.1501, PDHm >= 86.6629, PDHm <= 100.6986, CSm >= 28.17, CSm <= 51.2108, ACONTm >= 12.6008, ACONTm <= 40.3555, ICDHxm >= 0.0, ICDHxm <= 40.3559, ICDHym >= 0.0, ICDHym <= 30.8352, AKGDam >= 0.0, AKGDam <= 29.3331, AKGDbm >= 0.0, AKGDbm <= 29.3331, SUCD2_u6m >= 15.0997, SUCD2_u6m <= 40.2079, SUCD1m >= 15.0997, SUCD1m <= 40.2079, FUMm >= 17.1305, FUMm <= 41.8427, MDHm >= 28.17, MDHm <= 51.2108, ME1m >= -11.2237, ME1m <= -4.712, PC >= 0.0, PC <= 5.7862, ALDD2y >= 0.0, ALDD2y <= 1.1631, ACS - ACOAH >= 1.3221, ACS - ACOAH <= 3.5411, ICL >= 9.9873, ICL <= 20.8181, MALSp >= 9.9873, MALSp <= 20.8181, MDH >= 9.9871, MDH <= 21.7042, biomass_C + biomass_glucose + biomass_oil >= 8.6023, biomass_C + biomass_glucose + biomass_oil <= 10.3466, ATPM >= 457.2097, ATPM <= 813.9649'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update the media to minimal medium with glucose as the sole carbon source\n",
    "medium = model.medium\n",
    "medium['EX_glc_e'] = 100\n",
    "medium['EX_glyc_e'] = 0\n",
    "medium['EX_ocdcea_e'] = 0\n",
    "medium['EX_h2o_e'] = 10000\n",
    "medium['EX_h_e'] = 10000\n",
    "medium['EX_nh4_e'] = 10000\n",
    "medium['EX_o2_e'] = 10000\n",
    "medium['EX_pi_e'] = 10000\n",
    "medium['EX_so4_e'] = 10000\n",
    "medium['trehalose_c_tp'] = 0\n",
    "model.medium = medium\n",
    "\n",
    "# print the medium composition\n",
    "[print(model.medium[m], m) for m in model.medium]\n",
    "\n",
    "lower_bound_row = 'glucose_LB'\n",
    "upper_bound_row = 'glucose_UB'\n",
    "\n",
    "constraint_strings = []\n",
    "\n",
    "# loop over rows in the central flux dataframe\n",
    "for _, row in central_rxn_df.iterrows():  \n",
    "    # get the GSM reaction mapping for this reaction\n",
    "    reaction_ids = row['reaction_ids']\n",
    "    pathway = row['pathway']\n",
    "\n",
    "    # get the lower bound for this reaction\n",
    "    lower_bound = row[lower_bound_row]\n",
    "    upper_bound = row[upper_bound_row]\n",
    "\n",
    "    # determine if reaction should be included in the constraint string\n",
    "    reaction_ids_not_nan = not pd.isna(reaction_ids)\n",
    "    pathway_not_transport = pathway != 'transport'\n",
    "    bounds_are_not_nan = not pd.isna(lower_bound) and not pd.isna(upper_bound)\n",
    "    \n",
    "\n",
    "    if reaction_ids_not_nan and pathway_not_transport and bounds_are_not_nan:\n",
    "        print(f'13C-MFA bounds: {lower_bound} - {upper_bound}')\n",
    "\n",
    "        # make the constraint string\n",
    "        constraint_string = make_rxn_constraint_string(reaction_ids, lower_bound, upper_bound)\n",
    "\n",
    "        glucose_fba_solution = sd.fba(model, constraints=constraint_string, obj='biomass_glucose', obj_sense='maximize', pfba=1)\n",
    "\n",
    "        # add the constraint string to the list\n",
    "        constraint_strings.append(constraint_string)\n",
    "        print(constraint_string)\n",
    "        print(f\"max biomass flux: {glucose_fba_solution['biomass_glucose']}\")\n",
    "        print()\n",
    "\n",
    "\n",
    "\n",
    "full_constraint_string = ', '.join(constraint_strings)\n",
    "print(len(full_constraint_string))\n",
    "full_constraint_string "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate glucose GSM pFBA solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update the media to minimal medium with glucose as the sole carbon source\n",
    "medium = model.medium\n",
    "medium['EX_glc_e'] = 100\n",
    "medium['EX_glyc_e'] = 0\n",
    "medium['EX_ocdcea_e'] = 0\n",
    "medium['EX_h2o_e'] = 10000\n",
    "medium['EX_h_e'] = 10000\n",
    "medium['EX_nh4_e'] = 10000\n",
    "medium['EX_o2_e'] = 10000\n",
    "medium['EX_pi_e'] = 10000\n",
    "medium['EX_so4_e'] = 10000\n",
    "medium['trehalose_c_tp'] = 0\n",
    "model.medium = medium\n",
    "\n",
    "# print the medium composition\n",
    "[print(model.medium[m], m) for m in model.medium]\n",
    "\n",
    "temp_constraint_string = 'EX_glc_e >= -100.0, EX_glc_e <= -100.0, PGI >= 7.638, PGI <= 25.076, PFK - FBP >= 50.9378'\n",
    "glucose_fba_solution = sd.fba(model, constraints=full_constraint_string, obj='biomass_glucose', obj_sense='maximize', pfba=1)\n",
    "\n",
    "print()\n",
    "\n",
    "max_glucose_biomass_flux = glucose_fba_solution['biomass_glucose']\n",
    "PGI_flux = glucose_fba_solution['PGI']\n",
    "print(f'Maximum biomass flux: {max_glucose_biomass_flux}.')\n",
    "print(f'PGI flux: {PGI_flux}.')\n",
    "print(f'The number of active reactions in pFBA: {sum([abs(flux) > 0.1 for flux in glucose_fba_solution.fluxes.values()])}')\n",
    "\n",
    "# make a list of dictionaries with the reaction id, name, flux, and absolute flux\n",
    "reactions = []\n",
    "for reaction_id, flux in glucose_fba_solution.fluxes.items():\n",
    "\n",
    "  reactions.append({\n",
    "    'reaction_id': reaction_id,\n",
    "    'reaction_name': model.reactions.get_by_id(reaction_id).name,\n",
    "    'full_reaction': model.reactions.get_by_id(reaction_id).reaction,\n",
    "    'flux': flux,\n",
    "    'absolute_flux': abs(flux), # use for sorting, then drop\n",
    "  })\n",
    "\n",
    "# make a dataframe from the list of dictionaries\n",
    "glucose_gsm_fba_df = pd.DataFrame(reactions)\n",
    "\n",
    "# sort the dataframe by absolute flux\n",
    "glucose_gsm_fba_df = glucose_gsm_fba_df.sort_values(by=['absolute_flux'], ascending=False)\n",
    "\n",
    "# drop the absolute flux column\n",
    "glucose_gsm_fba_df = glucose_gsm_fba_df.drop(columns=['absolute_flux'])\n",
    "\n",
    "glucose_gsm_fba_df.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate glucose GSM pFBA FVA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run FVA for 90% of biomass production on the GSM\n",
    "biomass_fraction = 0.9\n",
    "glucose_fva_solution = sd.fva(\n",
    "  model, \n",
    "  constraints=f'EX_glc_e = -100.000, biomass_C >= {biomass_fraction * max_glucose_biomass_flux}',\n",
    ")\n",
    "\n",
    "# define a function to determine if a reaction is active\n",
    "def is_active(row):\n",
    "  return abs(row.maximum) > 0.1 or abs(row.minimum) > 0.1\n",
    "\n",
    "print(f'The number of active reactions in FVA: {sum([is_active(row) for _, row in glucose_fva_solution.iterrows()])}')\n",
    "\n",
    "# make a list of dictionaries with the reaction id, name, flux, and absolute flux\n",
    "fva_upper_bounds = []\n",
    "fva_lower_bounds = []\n",
    "\n",
    "# loop over the reactions in the GSM\n",
    "for _, row in glucose_gsm_fba_df.iterrows():\n",
    "  reaction_id = row.reaction_id\n",
    "\n",
    "  # get the upper and lower bounds from the FVA solution\n",
    "  upper_bound = glucose_fva_solution.loc[reaction_id, 'maximum']\n",
    "  lower_bound = glucose_fva_solution.loc[reaction_id, 'minimum']\n",
    "\n",
    "  fva_upper_bounds.append(upper_bound)\n",
    "  fva_lower_bounds.append(lower_bound)\n",
    "\n",
    "# add the upper and lower bounds to the dataframe\n",
    "glucose_gsm_fba_df['fva_upper_bound'] = fva_upper_bounds\n",
    "glucose_gsm_fba_df['fva_lower_bound'] = fva_lower_bounds\n",
    "\n",
    "# save the dataframe to a csv file\n",
    "glucose_gsm_fba_df.to_csv('../results/gsm_fluxes/glucose_gsm_13C_fba.csv', index=False)\n",
    "\n",
    "# display updated dataframe\n",
    "glucose_gsm_fba_df\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add glucose pFBA columns to 13C-MFA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the GSM flux predictions to the 13C-MFA dataframe\n",
    "central_rxn_df = add_flux_column_to_13c_flux_df(central_rxn_df, glucose_gsm_fba_df, 'glucose_pFBA_flux')\n",
    "\n",
    "# add the GSM flux predictions to the 13C-MFA dataframe\n",
    "central_rxn_df = add_fva_columns_to_13c_flux_df(central_rxn_df, glucose_gsm_fba_df, f'glucose_pFBA_{100*biomass_fraction}%')\n",
    "\n",
    "central_rxn_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Glycerol"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate glycerol GSM pFBA solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update the media to minimal medium with glycerol as the sole carbon source\n",
    "medium = model.medium\n",
    "medium['EX_glc_e'] = 0\n",
    "medium['EX_glyc_e'] = 100\n",
    "medium['EX_ocdcea_e'] = 0\n",
    "medium['EX_h2o_e'] = 10000\n",
    "medium['EX_h_e'] = 10000\n",
    "medium['EX_nh4_e'] = 10000\n",
    "medium['EX_o2_e'] = 10000\n",
    "medium['EX_pi_e'] = 10000\n",
    "medium['EX_so4_e'] = 10000\n",
    "medium['trehalose_c_tp'] = 0\n",
    "model.medium = medium\n",
    "\n",
    "# print the medium composition\n",
    "[print(model.medium[m], m) for m in model.medium]\n",
    "\n",
    "# run biomass-maximizing pFBA\n",
    "glycerol_fba_solution = sd.fba(model, constraints='EX_glyc_e = -100.000', obj='biomass_C', obj_sense='maximize', pfba=1)\n",
    "\n",
    "max_glycerol_biomass_flux = glycerol_fba_solution['biomass_C']\n",
    "print(f'Maximum biomass flux: {max_glycerol_biomass_flux}.')\n",
    "print(f'The number of active reactions in pFBA: {sum([abs(flux) > 0.1 for flux in glycerol_fba_solution.fluxes.values()])}')\n",
    "\n",
    "# make a list of dictionaries with the reaction id, name, flux, and absolute flux\n",
    "reactions = []\n",
    "for reaction_id, flux in glycerol_fba_solution.fluxes.items():\n",
    "\n",
    "  reactions.append({\n",
    "    'reaction_id': reaction_id,\n",
    "    'reaction_name': model.reactions.get_by_id(reaction_id).name,\n",
    "    'full_reaction': model.reactions.get_by_id(reaction_id).reaction,\n",
    "    'flux': flux,\n",
    "    'absolute_flux': abs(flux), # use for sorting, then drop\n",
    "  })\n",
    "\n",
    "# make a dataframe from the list of dictionaries\n",
    "glycerol_gsm_fba_df = pd.DataFrame(reactions)\n",
    "\n",
    "# sort the dataframe by absolute flux\n",
    "glycerol_gsm_fba_df = glycerol_gsm_fba_df.sort_values(by=['absolute_flux'], ascending=False)\n",
    "\n",
    "# drop the absolute flux column\n",
    "glycerol_gsm_fba_df = glycerol_gsm_fba_df.drop(columns=['absolute_flux'])\n",
    "\n",
    "glycerol_gsm_fba_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate glycerol GSM pFBA FVA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run FVA for 90% of biomass production on the GSM\n",
    "biomass_fraction = 0.9\n",
    "glycerol_fva_solution = sd.fva(\n",
    "  model, \n",
    "  constraints=f'EX_glyc_e = -100.000, biomass_C >= {biomass_fraction * max_glycerol_biomass_flux}',\n",
    ")\n",
    "\n",
    "# define a function to determine if a reaction is active\n",
    "def is_active(row):\n",
    "  return abs(row.maximum) > 0.1 or abs(row.minimum) > 0.1\n",
    "\n",
    "print(f'The number of active reactions in FVA: {sum([is_active(row) for _, row in glycerol_fva_solution.iterrows()])}')\n",
    "\n",
    "# make a list of dictionaries with the reaction id, name, flux, and absolute flux\n",
    "fva_upper_bounds = []\n",
    "fva_lower_bounds = []\n",
    "\n",
    "# loop over the reactions in the GSM\n",
    "for _, row in glycerol_gsm_fba_df.iterrows():\n",
    "  reaction_id = row.reaction_id\n",
    "\n",
    "  # get the upper and lower bounds from the FVA solution\n",
    "  upper_bound = glycerol_fva_solution.loc[reaction_id, 'maximum']\n",
    "  lower_bound = glycerol_fva_solution.loc[reaction_id, 'minimum']\n",
    "\n",
    "  fva_upper_bounds.append(upper_bound)\n",
    "  fva_lower_bounds.append(lower_bound)\n",
    "\n",
    "# add the upper and lower bounds to the dataframe\n",
    "glycerol_gsm_fba_df['fva_upper_bound'] = fva_upper_bounds\n",
    "glycerol_gsm_fba_df['fva_lower_bound'] = fva_lower_bounds\n",
    "\n",
    "# save the dataframe to a csv file\n",
    "glycerol_gsm_fba_df.to_csv('../results/gsm_fluxes/glycerol_gsm_13C_fba.csv', index=False)\n",
    "\n",
    "# display updated dataframe\n",
    "glycerol_gsm_fba_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add glycerol pFBA columns to 13C-MFA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the GSM flux predictions to the 13C-MFA dataframe\n",
    "central_rxn_df = add_flux_column_to_13c_flux_df(central_rxn_df, glycerol_gsm_fba_df, 'glycerol_pFBA_flux')\n",
    "\n",
    "# add the GSM flux predictions to the 13C-MFA dataframe\n",
    "central_rxn_df = add_fva_columns_to_13c_flux_df(central_rxn_df, glycerol_gsm_fba_df, f'glycerol_pFBA_{100*biomass_fraction}%')\n",
    "\n",
    "central_rxn_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Oleic Acid"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate oleic acid GSM pFBA solution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# update the media to minimal medium with oleic_acid as the sole carbon source\n",
    "medium = model.medium\n",
    "medium['EX_glc_e'] = 0\n",
    "medium['EX_glyc_e'] = 0\n",
    "medium['EX_ocdcea_e'] = 10 # this prevents overflow\n",
    "medium['EX_h2o_e'] = 10000\n",
    "medium['EX_h_e'] = 10000\n",
    "medium['EX_nh4_e'] = 10000\n",
    "medium['EX_o2_e'] = 10000\n",
    "medium['EX_pi_e'] = 10000\n",
    "medium['EX_so4_e'] = 10000\n",
    "medium['trehalose_c_tp'] = 0\n",
    "model.medium = medium\n",
    "\n",
    "# print the medium composition\n",
    "[print(model.medium[m], m) for m in model.medium]\n",
    "\n",
    "# run biomass-maximizing pFBA\n",
    "oleic_acid_fba_solution = sd.fba(model, constraints='EX_ocdcea_e = -10.000', obj='biomass_C', obj_sense='maximize', pfba=1)\n",
    "\n",
    "max_oleic_acid_biomass_flux = oleic_acid_fba_solution['biomass_C']\n",
    "print(f'Maximum biomass flux: {10 * max_oleic_acid_biomass_flux}.') # restore 100 input flux\n",
    "print(f'The number of active reactions in pFBA: {sum([abs(flux) > 0.1 for flux in oleic_acid_fba_solution.fluxes.values()])}')\n",
    "\n",
    "# make a list of dictionaries with the reaction id, name, flux, and absolute flux\n",
    "reactions = []\n",
    "for reaction_id, flux in oleic_acid_fba_solution.fluxes.items():\n",
    "\n",
    "  reactions.append({\n",
    "    'reaction_id': reaction_id,\n",
    "    'reaction_name': model.reactions.get_by_id(reaction_id).name,\n",
    "    'full_reaction': model.reactions.get_by_id(reaction_id).reaction,\n",
    "    'flux': 10 * flux, # restore 100 input flux\n",
    "    'absolute_flux': abs(flux), # use for sorting, then drop\n",
    "  })\n",
    "\n",
    "# make a dataframe from the list of dictionaries\n",
    "oleic_acid_gsm_fba_df = pd.DataFrame(reactions)\n",
    "\n",
    "# sort the dataframe by absolute flux\n",
    "oleic_acid_gsm_fba_df = oleic_acid_gsm_fba_df.sort_values(by=['absolute_flux'], ascending=False)\n",
    "\n",
    "# drop the absolute flux column\n",
    "oleic_acid_gsm_fba_df = oleic_acid_gsm_fba_df.drop(columns=['absolute_flux'])\n",
    "\n",
    "oleic_acid_gsm_fba_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculate oleic acid GSM pFBA FVA "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run FVA for 90% of biomass production on the GSM\n",
    "biomass_fraction = 0.9\n",
    "oleic_acid_fva_solution = sd.fva(\n",
    "  model, \n",
    "  constraints=f'EX_ocdcea_e = -10.000, biomass_C >= {biomass_fraction * max_oleic_acid_biomass_flux}',\n",
    ")\n",
    "\n",
    "# define a function to determine if a reaction is active\n",
    "def is_active(row):\n",
    "  return abs(row.maximum) > 0.1 or abs(row.minimum) > 0.1\n",
    "\n",
    "print(f'The number of active reactions in FVA: {sum([is_active(row) for _, row in oleic_acid_fva_solution.iterrows()])}')\n",
    "\n",
    "# make a list of dictionaries with the reaction id, name, flux, and absolute flux\n",
    "fva_upper_bounds = []\n",
    "fva_lower_bounds = []\n",
    "\n",
    "# loop over the reactions in the GSM\n",
    "for _, row in oleic_acid_gsm_fba_df.iterrows():\n",
    "  reaction_id = row.reaction_id\n",
    "\n",
    "  # get the upper and lower bounds from the FVA solution\n",
    "  upper_bound = 10 * oleic_acid_fva_solution.loc[reaction_id, 'maximum'] # restore 100 input flux\n",
    "  lower_bound = 10 * oleic_acid_fva_solution.loc[reaction_id, 'minimum'] # restore 100 input flux\n",
    "\n",
    "  fva_upper_bounds.append(upper_bound)\n",
    "  fva_lower_bounds.append(lower_bound)\n",
    "\n",
    "# add the upper and lower bounds to the dataframe\n",
    "oleic_acid_gsm_fba_df['fva_upper_bound'] = fva_upper_bounds\n",
    "oleic_acid_gsm_fba_df['fva_lower_bound'] = fva_lower_bounds\n",
    "\n",
    "# save the dataframe to a csv file\n",
    "oleic_acid_gsm_fba_df.to_csv('../results/gsm_fluxes/oleic_acid_gsm_13C_fba.csv', index=False)\n",
    "\n",
    "# display updated dataframe\n",
    "oleic_acid_gsm_fba_df\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add oleic acid pFBA columns to 13C-MFA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add the GSM flux predictions to the 13C-MFA dataframe\n",
    "central_rxn_df = add_flux_column_to_13c_flux_df(central_rxn_df, oleic_acid_gsm_fba_df, 'oleic_acid_pFBA_flux')\n",
    "\n",
    "# add the GSM flux predictions to the 13C-MFA dataframe\n",
    "central_rxn_df = add_fva_columns_to_13c_flux_df(central_rxn_df, oleic_acid_gsm_fba_df, f'oleic_acid_pFBA_{100*biomass_fraction}%')\n",
    "\n",
    "central_rxn_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save central flux data with pFBA data added"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the dataframe to a csv file\n",
    "central_rxn_df.to_csv('../results/central_fluxes/13C_pfba.csv', index=False, encoding='utf-8-sig')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
